{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_SmTogjeMD47"
      },
      "source": [
        "# 생물정보학 및 실습 1 - Free Project\n",
        "## LIN28A binding motif predictor 만들기 "
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### 환경 세팅 하기 "
      ],
      "metadata": {
        "id": "Il7jVuMN1DJ2"
      }
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2kWbS8JWMuHa",
        "outputId": "8c082600-7fad-4ba0-8797-47e1ba612911",
        "vscode": {
          "languageId": "python"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xQZGShwEMuHa"
      },
      "source": [
        " Bioconda 환경 설치\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_i5ZwVHUMuHb",
        "outputId": "7c1f03f2-13c9-4cb2-bba6-1cc0008fd2d4",
        "vscode": {
          "languageId": "python"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'colab-biolab'...\n",
            "remote: Enumerating objects: 69, done.\u001b[K\n",
            "remote: Counting objects: 100% (69/69), done.\u001b[K\n",
            "remote: Compressing objects: 100% (43/43), done.\u001b[K\n",
            "remote: Total 69 (delta 22), reused 54 (delta 13), pack-reused 0\u001b[K\n",
            "Unpacking objects: 100% (69/69), done.\n",
            "./\n",
            "./root/\n",
            "./root/.bin.priority/\n",
            "./root/.bin.priority/pip\n",
            "./root/.bin.priority/pip2\n",
            "./root/.bin.priority/pip3\n",
            "./root/.tmux.conf\n",
            "./root/.vimrc\n",
            "./root/.profile\n",
            "./root/.bashrc.biolab\n",
            "./root/.condarc\n",
            "--2022-06-08 08:13:02--  https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
            "Resolving repo.anaconda.com (repo.anaconda.com)... 104.16.131.3, 104.16.130.3, 2606:4700::6810:8303, ...\n",
            "Connecting to repo.anaconda.com (repo.anaconda.com)|104.16.131.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 76607678 (73M) [application/x-sh]\n",
            "Saving to: ‘miniconda3.sh’\n",
            "\n",
            "miniconda3.sh       100%[===================>]  73.06M   272MB/s    in 0.3s    \n",
            "\n",
            "2022-06-08 08:13:02 (272 MB/s) - ‘miniconda3.sh’ saved [76607678/76607678]\n",
            "\n",
            "PREFIX=/root/conda\n",
            "Unpacking payload ...\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\bdone\n",
            "Solving environment: - \b\b\\ \b\b| \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /root/conda\n",
            "\n",
            "  added / updated specs:\n",
            "    - _libgcc_mutex==0.1=main\n",
            "    - _openmp_mutex==4.5=1_gnu\n",
            "    - brotlipy==0.7.0=py39h27cfd23_1003\n",
            "    - ca-certificates==2022.3.29=h06a4308_1\n",
            "    - certifi==2021.10.8=py39h06a4308_2\n",
            "    - cffi==1.15.0=py39hd667e15_1\n",
            "    - charset-normalizer==2.0.4=pyhd3eb1b0_0\n",
            "    - colorama==0.4.4=pyhd3eb1b0_0\n",
            "    - conda-content-trust==0.1.1=pyhd3eb1b0_0\n",
            "    - conda-package-handling==1.8.1=py39h7f8727e_0\n",
            "    - conda==4.12.0=py39h06a4308_0\n",
            "    - cryptography==36.0.0=py39h9ce1e76_0\n",
            "    - idna==3.3=pyhd3eb1b0_0\n",
            "    - ld_impl_linux-64==2.35.1=h7274673_9\n",
            "    - libffi==3.3=he6710b0_2\n",
            "    - libgcc-ng==9.3.0=h5101ec6_17\n",
            "    - libgomp==9.3.0=h5101ec6_17\n",
            "    - libstdcxx-ng==9.3.0=hd4cf53a_17\n",
            "    - ncurses==6.3=h7f8727e_2\n",
            "    - openssl==1.1.1n=h7f8727e_0\n",
            "    - pip==21.2.4=py39h06a4308_0\n",
            "    - pycosat==0.6.3=py39h27cfd23_0\n",
            "    - pycparser==2.21=pyhd3eb1b0_0\n",
            "    - pyopenssl==22.0.0=pyhd3eb1b0_0\n",
            "    - pysocks==1.7.1=py39h06a4308_0\n",
            "    - python==3.9.12=h12debd9_0\n",
            "    - readline==8.1.2=h7f8727e_1\n",
            "    - requests==2.27.1=pyhd3eb1b0_0\n",
            "    - ruamel_yaml==0.15.100=py39h27cfd23_0\n",
            "    - setuptools==61.2.0=py39h06a4308_0\n",
            "    - six==1.16.0=pyhd3eb1b0_1\n",
            "    - sqlite==3.38.2=hc218d9a_0\n",
            "    - tk==8.6.11=h1ccaba5_0\n",
            "    - tqdm==4.63.0=pyhd3eb1b0_0\n",
            "    - tzdata==2022a=hda174b7_0\n",
            "    - urllib3==1.26.8=pyhd3eb1b0_0\n",
            "    - wheel==0.37.1=pyhd3eb1b0_0\n",
            "    - xz==5.2.5=h7b6447c_0\n",
            "    - yaml==0.2.5=h7b6447c_0\n",
            "    - zlib==1.2.12=h7f8727e_1\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main\n",
            "  _openmp_mutex      pkgs/main/linux-64::_openmp_mutex-4.5-1_gnu\n",
            "  brotlipy           pkgs/main/linux-64::brotlipy-0.7.0-py39h27cfd23_1003\n",
            "  ca-certificates    pkgs/main/linux-64::ca-certificates-2022.3.29-h06a4308_1\n",
            "  certifi            pkgs/main/linux-64::certifi-2021.10.8-py39h06a4308_2\n",
            "  cffi               pkgs/main/linux-64::cffi-1.15.0-py39hd667e15_1\n",
            "  charset-normalizer pkgs/main/noarch::charset-normalizer-2.0.4-pyhd3eb1b0_0\n",
            "  colorama           pkgs/main/noarch::colorama-0.4.4-pyhd3eb1b0_0\n",
            "  conda              pkgs/main/linux-64::conda-4.12.0-py39h06a4308_0\n",
            "  conda-content-tru~ pkgs/main/noarch::conda-content-trust-0.1.1-pyhd3eb1b0_0\n",
            "  conda-package-han~ pkgs/main/linux-64::conda-package-handling-1.8.1-py39h7f8727e_0\n",
            "  cryptography       pkgs/main/linux-64::cryptography-36.0.0-py39h9ce1e76_0\n",
            "  idna               pkgs/main/noarch::idna-3.3-pyhd3eb1b0_0\n",
            "  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.35.1-h7274673_9\n",
            "  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_2\n",
            "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.3.0-h5101ec6_17\n",
            "  libgomp            pkgs/main/linux-64::libgomp-9.3.0-h5101ec6_17\n",
            "  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.3.0-hd4cf53a_17\n",
            "  ncurses            pkgs/main/linux-64::ncurses-6.3-h7f8727e_2\n",
            "  openssl            pkgs/main/linux-64::openssl-1.1.1n-h7f8727e_0\n",
            "  pip                pkgs/main/linux-64::pip-21.2.4-py39h06a4308_0\n",
            "  pycosat            pkgs/main/linux-64::pycosat-0.6.3-py39h27cfd23_0\n",
            "  pycparser          pkgs/main/noarch::pycparser-2.21-pyhd3eb1b0_0\n",
            "  pyopenssl          pkgs/main/noarch::pyopenssl-22.0.0-pyhd3eb1b0_0\n",
            "  pysocks            pkgs/main/linux-64::pysocks-1.7.1-py39h06a4308_0\n",
            "  python             pkgs/main/linux-64::python-3.9.12-h12debd9_0\n",
            "  readline           pkgs/main/linux-64::readline-8.1.2-h7f8727e_1\n",
            "  requests           pkgs/main/noarch::requests-2.27.1-pyhd3eb1b0_0\n",
            "  ruamel_yaml        pkgs/main/linux-64::ruamel_yaml-0.15.100-py39h27cfd23_0\n",
            "  setuptools         pkgs/main/linux-64::setuptools-61.2.0-py39h06a4308_0\n",
            "  six                pkgs/main/noarch::six-1.16.0-pyhd3eb1b0_1\n",
            "  sqlite             pkgs/main/linux-64::sqlite-3.38.2-hc218d9a_0\n",
            "  tk                 pkgs/main/linux-64::tk-8.6.11-h1ccaba5_0\n",
            "  tqdm               pkgs/main/noarch::tqdm-4.63.0-pyhd3eb1b0_0\n",
            "  tzdata             pkgs/main/noarch::tzdata-2022a-hda174b7_0\n",
            "  urllib3            pkgs/main/noarch::urllib3-1.26.8-pyhd3eb1b0_0\n",
            "  wheel              pkgs/main/noarch::wheel-0.37.1-pyhd3eb1b0_0\n",
            "  xz                 pkgs/main/linux-64::xz-5.2.5-h7b6447c_0\n",
            "  yaml               pkgs/main/linux-64::yaml-0.2.5-h7b6447c_0\n",
            "  zlib               pkgs/main/linux-64::zlib-1.2.12-h7f8727e_1\n",
            "\n",
            "\n",
            "Preparing transaction: - \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Executing transaction: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "installation finished.\n",
            "WARNING:\n",
            "    You currently have a PYTHONPATH environment variable set. This may cause\n",
            "    unexpected behavior when running the Python interpreter in Miniconda3.\n",
            "    For best results, please verify that your PYTHONPATH only points to\n",
            "    directories of packages that are compatible with the Python interpreter\n",
            "    in Miniconda3: /root/conda\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Solving environment: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.12.0\n",
            "  latest version: 4.13.0\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /root/conda/envs/lab\n",
            "\n",
            "  added / updated specs:\n",
            "    - bedtools\n",
            "    - biopython\n",
            "    - bwa\n",
            "    - cython\n",
            "    - matplotlib\n",
            "    - numpy\n",
            "    - pandas\n",
            "    - pip\n",
            "    - pysam\n",
            "    - samtools\n",
            "    - scipy\n",
            "    - star\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge\n",
            "    _openmp_mutex-4.5          |            2_gnu          23 KB  conda-forge\n",
            "    alsa-lib-1.2.3.2           |       h166bdaf_0         554 KB  conda-forge\n",
            "    attr-2.5.1                 |       h166bdaf_0          69 KB  conda-forge\n",
            "    bedtools-2.30.0            |       h468198e_3        15.5 MB  bioconda\n",
            "    biopython-1.79             |   py39h3811e60_1         2.6 MB  conda-forge\n",
            "    brotli-1.0.9               |       h166bdaf_7          18 KB  conda-forge\n",
            "    brotli-bin-1.0.9           |       h166bdaf_7          19 KB  conda-forge\n",
            "    bwa-0.7.17                 |       h7132678_9         187 KB  bioconda\n",
            "    bzip2-1.0.8                |       h7f98852_4         484 KB  conda-forge\n",
            "    c-ares-1.18.1              |       h7f98852_0         113 KB  conda-forge\n",
            "    ca-certificates-2022.5.18.1|       ha878542_0         144 KB  conda-forge\n",
            "    certifi-2022.5.18.1        |   py39hf3d152e_0         150 KB  conda-forge\n",
            "    cycler-0.11.0              |     pyhd8ed1ab_0          10 KB  conda-forge\n",
            "    cython-0.29.30             |   py39h5a03fae_0         2.1 MB  conda-forge\n",
            "    dbus-1.13.6                |       h5008d03_3         604 KB  conda-forge\n",
            "    expat-2.4.8                |       h27087fc_0         187 KB  conda-forge\n",
            "    fftw-3.3.10                |nompi_h77c792f_102         6.4 MB  conda-forge\n",
            "    font-ttf-dejavu-sans-mono-2.37|       hab24e00_0         388 KB  conda-forge\n",
            "    font-ttf-inconsolata-3.000 |       h77eed37_0          94 KB  conda-forge\n",
            "    font-ttf-source-code-pro-2.038|       h77eed37_0         684 KB  conda-forge\n",
            "    font-ttf-ubuntu-0.83       |       hab24e00_0         1.9 MB  conda-forge\n",
            "    fontconfig-2.14.0          |       h8e229c2_0         305 KB  conda-forge\n",
            "    fonts-conda-ecosystem-1    |                0           4 KB  conda-forge\n",
            "    fonts-conda-forge-1        |                0           4 KB  conda-forge\n",
            "    fonttools-4.33.3           |   py39hb9d737c_0         1.6 MB  conda-forge\n",
            "    freetype-2.10.4            |       h0708190_1         890 KB  conda-forge\n",
            "    gettext-0.19.8.1           |    h73d1719_1008         3.6 MB  conda-forge\n",
            "    giflib-5.2.1               |       h36c2ea0_2          77 KB  conda-forge\n",
            "    gst-plugins-base-1.20.2    |       hcf0ee16_0         2.8 MB  conda-forge\n",
            "    gstreamer-1.20.2           |       hd4edc92_1         2.0 MB  conda-forge\n",
            "    htslib-1.15.1              |       h9753748_0         2.1 MB  bioconda\n",
            "    icu-70.1                   |       h27087fc_0        13.5 MB  conda-forge\n",
            "    jack-1.9.18                |    hfd4fe87_1001         640 KB  conda-forge\n",
            "    jpeg-9e                    |       h166bdaf_1         268 KB  conda-forge\n",
            "    keyutils-1.6.1             |       h166bdaf_0         115 KB  conda-forge\n",
            "    kiwisolver-1.4.2           |   py39hf939315_1          75 KB  conda-forge\n",
            "    krb5-1.19.3                |       h3790be6_0         1.4 MB  conda-forge\n",
            "    lcms2-2.12                 |       hddcbb42_0         443 KB  conda-forge\n",
            "    ld_impl_linux-64-2.36.1    |       hea4e1c9_2         667 KB  conda-forge\n",
            "    lerc-3.0                   |       h9c3ff4c_0         216 KB  conda-forge\n",
            "    libblas-3.9.0              |15_linux64_openblas          12 KB  conda-forge\n",
            "    libbrotlicommon-1.0.9      |       h166bdaf_7          65 KB  conda-forge\n",
            "    libbrotlidec-1.0.9         |       h166bdaf_7          33 KB  conda-forge\n",
            "    libbrotlienc-1.0.9         |       h166bdaf_7         287 KB  conda-forge\n",
            "    libcap-2.64                |       ha37c62d_0          96 KB  conda-forge\n",
            "    libcblas-3.9.0             |15_linux64_openblas          12 KB  conda-forge\n",
            "    libclang-14.0.4            |default_h2e3cab8_0         127 KB  conda-forge\n",
            "    libclang13-14.0.4          |default_h3a83d3e_0        10.6 MB  conda-forge\n",
            "    libcups-2.3.3              |       hf5a7f15_1         4.6 MB  conda-forge\n",
            "    libcurl-7.83.1             |       h7bff187_0         342 KB  conda-forge\n",
            "    libdb-6.2.32               |       h9c3ff4c_0        23.3 MB  conda-forge\n",
            "    libdeflate-1.10            |       h7f98852_0          77 KB  conda-forge\n",
            "    libedit-3.1.20191231       |       he28a2e2_2         121 KB  conda-forge\n",
            "    libev-4.33                 |       h516909a_1         104 KB  conda-forge\n",
            "    libevent-2.1.10            |       h9b69904_4         1.1 MB  conda-forge\n",
            "    libffi-3.4.2               |       h7f98852_5          57 KB  conda-forge\n",
            "    libflac-1.3.4              |       h27087fc_0         474 KB  conda-forge\n",
            "    libgcc-ng-12.1.0           |      h8d9b700_16         940 KB  conda-forge\n",
            "    libgfortran-ng-12.1.0      |      h69a702a_16          23 KB  conda-forge\n",
            "    libgfortran5-12.1.0        |      hdcd56e2_16         1.8 MB  conda-forge\n",
            "    libglib-2.70.2             |       h174f98d_4         3.1 MB  conda-forge\n",
            "    libgomp-12.1.0             |      h8d9b700_16         459 KB  conda-forge\n",
            "    libiconv-1.16              |       h516909a_0         1.4 MB  conda-forge\n",
            "    liblapack-3.9.0            |15_linux64_openblas          12 KB  conda-forge\n",
            "    libllvm14-14.0.4           |       he0ac6c6_0        35.2 MB  conda-forge\n",
            "    libnghttp2-1.47.0          |       h727a467_0         808 KB  conda-forge\n",
            "    libnsl-2.0.0               |       h7f98852_0          31 KB  conda-forge\n",
            "    libogg-1.3.4               |       h7f98852_1         206 KB  conda-forge\n",
            "    libopenblas-0.3.20         |pthreads_h78a6416_0        10.1 MB  conda-forge\n",
            "    libopus-1.3.1              |       h7f98852_1         255 KB  conda-forge\n",
            "    libpng-1.6.37              |       h21135ba_2         306 KB  conda-forge\n",
            "    libpq-14.3                 |       hd77ab85_0         3.0 MB  conda-forge\n",
            "    libsndfile-1.0.31          |       h9c3ff4c_1         602 KB  conda-forge\n",
            "    libssh2-1.10.0             |       ha56f1ee_2         233 KB  conda-forge\n",
            "    libstdcxx-ng-12.1.0        |      ha89aaad_16         4.3 MB  conda-forge\n",
            "    libtiff-4.4.0              |       h0fcbabc_0         606 KB  conda-forge\n",
            "    libtool-2.4.6              |    h9c3ff4c_1008         511 KB  conda-forge\n",
            "    libudev1-249               |       h166bdaf_3         109 KB  conda-forge\n",
            "    libuuid-2.32.1             |    h7f98852_1000          28 KB  conda-forge\n",
            "    libvorbis-1.3.7            |       h9c3ff4c_0         280 KB  conda-forge\n",
            "    libwebp-1.2.2              |       h3452ae3_0          85 KB  conda-forge\n",
            "    libwebp-base-1.2.2         |       h7f98852_1         824 KB  conda-forge\n",
            "    libxcb-1.13                |    h7f98852_1004         391 KB  conda-forge\n",
            "    libxkbcommon-1.0.3         |       he3ba5ed_0         581 KB  conda-forge\n",
            "    libxml2-2.9.14             |       h22db469_0         770 KB  conda-forge\n",
            "    libzlib-1.2.12             |       h166bdaf_0          62 KB  conda-forge\n",
            "    lz4-c-1.9.3                |       h9c3ff4c_1         179 KB  conda-forge\n",
            "    matplotlib-3.5.2           |   py39hf3d152e_0           6 KB  conda-forge\n",
            "    matplotlib-base-3.5.2      |   py39h700656a_0         7.3 MB  conda-forge\n",
            "    munkres-1.1.4              |     pyh9f0ad1d_0          12 KB  conda-forge\n",
            "    mysql-common-8.0.29        |       haf5c9bc_1         1.8 MB  conda-forge\n",
            "    mysql-libs-8.0.29          |       h28c427c_1         1.9 MB  conda-forge\n",
            "    ncurses-6.3                |       h27087fc_1        1002 KB  conda-forge\n",
            "    nspr-4.32                  |       h9c3ff4c_1         233 KB  conda-forge\n",
            "    nss-3.78                   |       h2350873_0         2.1 MB  conda-forge\n",
            "    numpy-1.22.4               |   py39hc58783e_0         6.8 MB  conda-forge\n",
            "    openjpeg-2.4.0             |       hb52868f_1         444 KB  conda-forge\n",
            "    openssl-1.1.1o             |       h166bdaf_0         2.1 MB  conda-forge\n",
            "    packaging-21.3             |     pyhd8ed1ab_0          36 KB  conda-forge\n",
            "    pandas-1.4.2               |   py39h1832856_2        12.5 MB  conda-forge\n",
            "    pcre-8.45                  |       h9c3ff4c_0         253 KB  conda-forge\n",
            "    perl-5.32.1                | 2_h7f98852_perl5        14.4 MB  conda-forge\n",
            "    pillow-9.1.1               |   py39hae2aec6_1        44.5 MB  conda-forge\n",
            "    pip-22.1.2                 |     pyhd8ed1ab_0         1.5 MB  conda-forge\n",
            "    pthread-stubs-0.4          |    h36c2ea0_1001           5 KB  conda-forge\n",
            "    pulseaudio-14.0            |       hbc9ff1d_7         1.7 MB  conda-forge\n",
            "    pyparsing-3.0.9            |     pyhd8ed1ab_0          79 KB  conda-forge\n",
            "    pyqt-5.15.4                |   py39h18e9c17_1         6.1 MB  conda-forge\n",
            "    pyqt5-sip-12.9.0           |   py39h5a03fae_1          84 KB  conda-forge\n",
            "    pysam-0.19.1               |   py39h5030a8b_0         2.7 MB  bioconda\n",
            "    python-3.9.13              |h9a8a25e_0_cpython        26.7 MB  conda-forge\n",
            "    python-dateutil-2.8.2      |     pyhd8ed1ab_0         240 KB  conda-forge\n",
            "    python_abi-3.9             |           2_cp39           4 KB  conda-forge\n",
            "    pytz-2022.1                |     pyhd8ed1ab_0         242 KB  conda-forge\n",
            "    qt-main-5.15.3             |       hf97cb25_2        62.0 MB  conda-forge\n",
            "    readline-8.1               |       h46c0cb4_0         295 KB  conda-forge\n",
            "    samtools-1.15.1            |       h1170115_0         393 KB  bioconda\n",
            "    scipy-1.8.1                |   py39he49c0e8_0        24.9 MB  conda-forge\n",
            "    setuptools-62.3.3          |   py39hf3d152e_0         1.3 MB  conda-forge\n",
            "    sip-6.5.1                  |   py39he80948d_2         389 KB  conda-forge\n",
            "    six-1.16.0                 |     pyh6c4a22f_0          14 KB  conda-forge\n",
            "    sqlite-3.38.5              |       h4ff8645_0         1.5 MB  conda-forge\n",
            "    star-2.7.10a               |       h9ee0642_0         4.9 MB  bioconda\n",
            "    tk-8.6.12                  |       h27826a3_0         3.3 MB  conda-forge\n",
            "    toml-0.10.2                |     pyhd8ed1ab_0          18 KB  conda-forge\n",
            "    tornado-6.1                |   py39hb9d737c_3         649 KB  conda-forge\n",
            "    tzdata-2022a               |       h191b570_0         121 KB  conda-forge\n",
            "    unicodedata2-14.0.0        |   py39hb9d737c_1         498 KB  conda-forge\n",
            "    wheel-0.37.1               |     pyhd8ed1ab_0          31 KB  conda-forge\n",
            "    xorg-libxau-1.0.9          |       h7f98852_0          13 KB  conda-forge\n",
            "    xorg-libxdmcp-1.1.3        |       h7f98852_0          19 KB  conda-forge\n",
            "    xz-5.2.5                   |       h516909a_1         343 KB  conda-forge\n",
            "    zlib-1.2.12                |       h166bdaf_0          90 KB  conda-forge\n",
            "    zstd-1.5.2                 |       h8a70e8d_1         452 KB  conda-forge\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:       407.1 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
            "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-2_gnu\n",
            "  alsa-lib           conda-forge/linux-64::alsa-lib-1.2.3.2-h166bdaf_0\n",
            "  attr               conda-forge/linux-64::attr-2.5.1-h166bdaf_0\n",
            "  bedtools           bioconda/linux-64::bedtools-2.30.0-h468198e_3\n",
            "  biopython          conda-forge/linux-64::biopython-1.79-py39h3811e60_1\n",
            "  brotli             conda-forge/linux-64::brotli-1.0.9-h166bdaf_7\n",
            "  brotli-bin         conda-forge/linux-64::brotli-bin-1.0.9-h166bdaf_7\n",
            "  bwa                bioconda/linux-64::bwa-0.7.17-h7132678_9\n",
            "  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4\n",
            "  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0\n",
            "  ca-certificates    conda-forge/linux-64::ca-certificates-2022.5.18.1-ha878542_0\n",
            "  certifi            conda-forge/linux-64::certifi-2022.5.18.1-py39hf3d152e_0\n",
            "  cycler             conda-forge/noarch::cycler-0.11.0-pyhd8ed1ab_0\n",
            "  cython             conda-forge/linux-64::cython-0.29.30-py39h5a03fae_0\n",
            "  dbus               conda-forge/linux-64::dbus-1.13.6-h5008d03_3\n",
            "  expat              conda-forge/linux-64::expat-2.4.8-h27087fc_0\n",
            "  fftw               conda-forge/linux-64::fftw-3.3.10-nompi_h77c792f_102\n",
            "  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0\n",
            "  font-ttf-inconsol~ conda-forge/noarch::font-ttf-inconsolata-3.000-h77eed37_0\n",
            "  font-ttf-source-c~ conda-forge/noarch::font-ttf-source-code-pro-2.038-h77eed37_0\n",
            "  font-ttf-ubuntu    conda-forge/noarch::font-ttf-ubuntu-0.83-hab24e00_0\n",
            "  fontconfig         conda-forge/linux-64::fontconfig-2.14.0-h8e229c2_0\n",
            "  fonts-conda-ecosy~ conda-forge/noarch::fonts-conda-ecosystem-1-0\n",
            "  fonts-conda-forge  conda-forge/noarch::fonts-conda-forge-1-0\n",
            "  fonttools          conda-forge/linux-64::fonttools-4.33.3-py39hb9d737c_0\n",
            "  freetype           conda-forge/linux-64::freetype-2.10.4-h0708190_1\n",
            "  gettext            conda-forge/linux-64::gettext-0.19.8.1-h73d1719_1008\n",
            "  giflib             conda-forge/linux-64::giflib-5.2.1-h36c2ea0_2\n",
            "  gst-plugins-base   conda-forge/linux-64::gst-plugins-base-1.20.2-hcf0ee16_0\n",
            "  gstreamer          conda-forge/linux-64::gstreamer-1.20.2-hd4edc92_1\n",
            "  htslib             bioconda/linux-64::htslib-1.15.1-h9753748_0\n",
            "  icu                conda-forge/linux-64::icu-70.1-h27087fc_0\n",
            "  jack               conda-forge/linux-64::jack-1.9.18-hfd4fe87_1001\n",
            "  jpeg               conda-forge/linux-64::jpeg-9e-h166bdaf_1\n",
            "  keyutils           conda-forge/linux-64::keyutils-1.6.1-h166bdaf_0\n",
            "  kiwisolver         conda-forge/linux-64::kiwisolver-1.4.2-py39hf939315_1\n",
            "  krb5               conda-forge/linux-64::krb5-1.19.3-h3790be6_0\n",
            "  lcms2              conda-forge/linux-64::lcms2-2.12-hddcbb42_0\n",
            "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.36.1-hea4e1c9_2\n",
            "  lerc               conda-forge/linux-64::lerc-3.0-h9c3ff4c_0\n",
            "  libblas            conda-forge/linux-64::libblas-3.9.0-15_linux64_openblas\n",
            "  libbrotlicommon    conda-forge/linux-64::libbrotlicommon-1.0.9-h166bdaf_7\n",
            "  libbrotlidec       conda-forge/linux-64::libbrotlidec-1.0.9-h166bdaf_7\n",
            "  libbrotlienc       conda-forge/linux-64::libbrotlienc-1.0.9-h166bdaf_7\n",
            "  libcap             conda-forge/linux-64::libcap-2.64-ha37c62d_0\n",
            "  libcblas           conda-forge/linux-64::libcblas-3.9.0-15_linux64_openblas\n",
            "  libclang           conda-forge/linux-64::libclang-14.0.4-default_h2e3cab8_0\n",
            "  libclang13         conda-forge/linux-64::libclang13-14.0.4-default_h3a83d3e_0\n",
            "  libcups            conda-forge/linux-64::libcups-2.3.3-hf5a7f15_1\n",
            "  libcurl            conda-forge/linux-64::libcurl-7.83.1-h7bff187_0\n",
            "  libdb              conda-forge/linux-64::libdb-6.2.32-h9c3ff4c_0\n",
            "  libdeflate         conda-forge/linux-64::libdeflate-1.10-h7f98852_0\n",
            "  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2\n",
            "  libev              conda-forge/linux-64::libev-4.33-h516909a_1\n",
            "  libevent           conda-forge/linux-64::libevent-2.1.10-h9b69904_4\n",
            "  libffi             conda-forge/linux-64::libffi-3.4.2-h7f98852_5\n",
            "  libflac            conda-forge/linux-64::libflac-1.3.4-h27087fc_0\n",
            "  libgcc-ng          conda-forge/linux-64::libgcc-ng-12.1.0-h8d9b700_16\n",
            "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-12.1.0-h69a702a_16\n",
            "  libgfortran5       conda-forge/linux-64::libgfortran5-12.1.0-hdcd56e2_16\n",
            "  libglib            conda-forge/linux-64::libglib-2.70.2-h174f98d_4\n",
            "  libgomp            conda-forge/linux-64::libgomp-12.1.0-h8d9b700_16\n",
            "  libiconv           conda-forge/linux-64::libiconv-1.16-h516909a_0\n",
            "  liblapack          conda-forge/linux-64::liblapack-3.9.0-15_linux64_openblas\n",
            "  libllvm14          conda-forge/linux-64::libllvm14-14.0.4-he0ac6c6_0\n",
            "  libnghttp2         conda-forge/linux-64::libnghttp2-1.47.0-h727a467_0\n",
            "  libnsl             conda-forge/linux-64::libnsl-2.0.0-h7f98852_0\n",
            "  libogg             conda-forge/linux-64::libogg-1.3.4-h7f98852_1\n",
            "  libopenblas        conda-forge/linux-64::libopenblas-0.3.20-pthreads_h78a6416_0\n",
            "  libopus            conda-forge/linux-64::libopus-1.3.1-h7f98852_1\n",
            "  libpng             conda-forge/linux-64::libpng-1.6.37-h21135ba_2\n",
            "  libpq              conda-forge/linux-64::libpq-14.3-hd77ab85_0\n",
            "  libsndfile         conda-forge/linux-64::libsndfile-1.0.31-h9c3ff4c_1\n",
            "  libssh2            conda-forge/linux-64::libssh2-1.10.0-ha56f1ee_2\n",
            "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-12.1.0-ha89aaad_16\n",
            "  libtiff            conda-forge/linux-64::libtiff-4.4.0-h0fcbabc_0\n",
            "  libtool            conda-forge/linux-64::libtool-2.4.6-h9c3ff4c_1008\n",
            "  libudev1           conda-forge/linux-64::libudev1-249-h166bdaf_3\n",
            "  libuuid            conda-forge/linux-64::libuuid-2.32.1-h7f98852_1000\n",
            "  libvorbis          conda-forge/linux-64::libvorbis-1.3.7-h9c3ff4c_0\n",
            "  libwebp            conda-forge/linux-64::libwebp-1.2.2-h3452ae3_0\n",
            "  libwebp-base       conda-forge/linux-64::libwebp-base-1.2.2-h7f98852_1\n",
            "  libxcb             conda-forge/linux-64::libxcb-1.13-h7f98852_1004\n",
            "  libxkbcommon       conda-forge/linux-64::libxkbcommon-1.0.3-he3ba5ed_0\n",
            "  libxml2            conda-forge/linux-64::libxml2-2.9.14-h22db469_0\n",
            "  libzlib            conda-forge/linux-64::libzlib-1.2.12-h166bdaf_0\n",
            "  lz4-c              conda-forge/linux-64::lz4-c-1.9.3-h9c3ff4c_1\n",
            "  matplotlib         conda-forge/linux-64::matplotlib-3.5.2-py39hf3d152e_0\n",
            "  matplotlib-base    conda-forge/linux-64::matplotlib-base-3.5.2-py39h700656a_0\n",
            "  munkres            conda-forge/noarch::munkres-1.1.4-pyh9f0ad1d_0\n",
            "  mysql-common       conda-forge/linux-64::mysql-common-8.0.29-haf5c9bc_1\n",
            "  mysql-libs         conda-forge/linux-64::mysql-libs-8.0.29-h28c427c_1\n",
            "  ncurses            conda-forge/linux-64::ncurses-6.3-h27087fc_1\n",
            "  nspr               conda-forge/linux-64::nspr-4.32-h9c3ff4c_1\n",
            "  nss                conda-forge/linux-64::nss-3.78-h2350873_0\n",
            "  numpy              conda-forge/linux-64::numpy-1.22.4-py39hc58783e_0\n",
            "  openjpeg           conda-forge/linux-64::openjpeg-2.4.0-hb52868f_1\n",
            "  openssl            conda-forge/linux-64::openssl-1.1.1o-h166bdaf_0\n",
            "  packaging          conda-forge/noarch::packaging-21.3-pyhd8ed1ab_0\n",
            "  pandas             conda-forge/linux-64::pandas-1.4.2-py39h1832856_2\n",
            "  pcre               conda-forge/linux-64::pcre-8.45-h9c3ff4c_0\n",
            "  perl               conda-forge/linux-64::perl-5.32.1-2_h7f98852_perl5\n",
            "  pillow             conda-forge/linux-64::pillow-9.1.1-py39hae2aec6_1\n",
            "  pip                conda-forge/noarch::pip-22.1.2-pyhd8ed1ab_0\n",
            "  pthread-stubs      conda-forge/linux-64::pthread-stubs-0.4-h36c2ea0_1001\n",
            "  pulseaudio         conda-forge/linux-64::pulseaudio-14.0-hbc9ff1d_7\n",
            "  pyparsing          conda-forge/noarch::pyparsing-3.0.9-pyhd8ed1ab_0\n",
            "  pyqt               conda-forge/linux-64::pyqt-5.15.4-py39h18e9c17_1\n",
            "  pyqt5-sip          conda-forge/linux-64::pyqt5-sip-12.9.0-py39h5a03fae_1\n",
            "  pysam              bioconda/linux-64::pysam-0.19.1-py39h5030a8b_0\n",
            "  python             conda-forge/linux-64::python-3.9.13-h9a8a25e_0_cpython\n",
            "  python-dateutil    conda-forge/noarch::python-dateutil-2.8.2-pyhd8ed1ab_0\n",
            "  python_abi         conda-forge/linux-64::python_abi-3.9-2_cp39\n",
            "  pytz               conda-forge/noarch::pytz-2022.1-pyhd8ed1ab_0\n",
            "  qt-main            conda-forge/linux-64::qt-main-5.15.3-hf97cb25_2\n",
            "  readline           conda-forge/linux-64::readline-8.1-h46c0cb4_0\n",
            "  samtools           bioconda/linux-64::samtools-1.15.1-h1170115_0\n",
            "  scipy              conda-forge/linux-64::scipy-1.8.1-py39he49c0e8_0\n",
            "  setuptools         conda-forge/linux-64::setuptools-62.3.3-py39hf3d152e_0\n",
            "  sip                conda-forge/linux-64::sip-6.5.1-py39he80948d_2\n",
            "  six                conda-forge/noarch::six-1.16.0-pyh6c4a22f_0\n",
            "  sqlite             conda-forge/linux-64::sqlite-3.38.5-h4ff8645_0\n",
            "  star               bioconda/linux-64::star-2.7.10a-h9ee0642_0\n",
            "  tk                 conda-forge/linux-64::tk-8.6.12-h27826a3_0\n",
            "  toml               conda-forge/noarch::toml-0.10.2-pyhd8ed1ab_0\n",
            "  tornado            conda-forge/linux-64::tornado-6.1-py39hb9d737c_3\n",
            "  tzdata             conda-forge/noarch::tzdata-2022a-h191b570_0\n",
            "  unicodedata2       conda-forge/linux-64::unicodedata2-14.0.0-py39hb9d737c_1\n",
            "  wheel              conda-forge/noarch::wheel-0.37.1-pyhd8ed1ab_0\n",
            "  xorg-libxau        conda-forge/linux-64::xorg-libxau-1.0.9-h7f98852_0\n",
            "  xorg-libxdmcp      conda-forge/linux-64::xorg-libxdmcp-1.1.3-h7f98852_0\n",
            "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
            "  zlib               conda-forge/linux-64::zlib-1.2.12-h166bdaf_0\n",
            "  zstd               conda-forge/linux-64::zstd-1.5.2-h8a70e8d_1\n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "libdb-6.2.32         | 23.3 MB   | : 100% 1.0/1 [00:05<00:00,  5.61s/it]               \n",
            "libdeflate-1.10      | 77 KB     | : 100% 1.0/1 [00:00<00:00, 23.47it/s]\n",
            "liblapack-3.9.0      | 12 KB     | : 100% 1.0/1 [00:00<00:00, 30.73it/s]\n",
            "libxml2-2.9.14       | 770 KB    | : 100% 1.0/1 [00:00<00:00,  5.19it/s]\n",
            "libzlib-1.2.12       | 62 KB     | : 100% 1.0/1 [00:00<00:00, 21.27it/s]\n",
            "gst-plugins-base-1.2 | 2.8 MB    | : 100% 1.0/1 [00:00<00:00,  1.31it/s]\n",
            "libcblas-3.9.0       | 12 KB     | : 100% 1.0/1 [00:00<00:00, 33.82it/s]\n",
            "bzip2-1.0.8          | 484 KB    | : 100% 1.0/1 [00:00<00:00,  6.72it/s]\n",
            "readline-8.1         | 295 KB    | : 100% 1.0/1 [00:00<00:00, 10.88it/s]\n",
            "libsndfile-1.0.31    | 602 KB    | : 100% 1.0/1 [00:00<00:00,  5.48it/s]\n",
            "font-ttf-inconsolata | 94 KB     | : 100% 1.0/1 [00:00<00:00, 23.24it/s]\n",
            "setuptools-62.3.3    | 1.3 MB    | : 100% 1.0/1 [00:00<00:00,  2.20it/s]\n",
            "libwebp-base-1.2.2   | 824 KB    | : 100% 1.0/1 [00:00<00:00,  5.31it/s]\n",
            "samtools-1.15.1      | 393 KB    | : 100% 1.0/1 [00:00<00:00,  9.22it/s]\n",
            "dbus-1.13.6          | 604 KB    | : 100% 1.0/1 [00:00<00:00,  6.57it/s]\n",
            "jack-1.9.18          | 640 KB    | : 100% 1.0/1 [00:00<00:00,  5.86it/s]\n",
            "mysql-libs-8.0.29    | 1.9 MB    | : 100% 1.0/1 [00:00<00:00,  2.67it/s]\n",
            "biopython-1.79       | 2.6 MB    | : 100% 1.0/1 [00:00<00:00,  1.28it/s]\n",
            "attr-2.5.1           | 69 KB     | : 100% 1.0/1 [00:00<00:00, 17.65it/s]\n",
            "pyqt-5.15.4          | 6.1 MB    | : 100% 1.0/1 [00:01<00:00,  1.64s/it]               \n",
            "pthread-stubs-0.4    | 5 KB      | : 100% 1.0/1 [00:00<00:00, 33.20it/s]\n",
            "libevent-2.1.10      | 1.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.97it/s]\n",
            "libglib-2.70.2       | 3.1 MB    | : 100% 1.0/1 [00:00<00:00,  1.31it/s]\n",
            "fftw-3.3.10          | 6.4 MB    | : 100% 1.0/1 [00:01<00:00,  1.39s/it]               \n",
            "libvorbis-1.3.7      | 280 KB    | : 100% 1.0/1 [00:00<00:00, 12.24it/s]\n",
            "sqlite-3.38.5        | 1.5 MB    | : 100% 1.0/1 [00:00<00:00,  3.50it/s]\n",
            "libstdcxx-ng-12.1.0  | 4.3 MB    | : 100% 1.0/1 [00:00<00:00,  1.38it/s]\n",
            "libtiff-4.4.0        | 606 KB    | : 100% 1.0/1 [00:00<00:00,  6.46it/s]\n",
            "munkres-1.1.4        | 12 KB     | : 100% 1.0/1 [00:00<00:00, 32.13it/s]\n",
            "libblas-3.9.0        | 12 KB     | : 100% 1.0/1 [00:00<00:00, 31.87it/s]\n",
            "gstreamer-1.20.2     | 2.0 MB    | : 100% 1.0/1 [00:00<00:00,  1.87it/s]\n",
            "libssh2-1.10.0       | 233 KB    | : 100% 1.0/1 [00:00<00:00, 11.88it/s]\n",
            "htslib-1.15.1        | 2.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.50it/s]\n",
            "expat-2.4.8          | 187 KB    | : 100% 1.0/1 [00:00<00:00, 14.06it/s]\n",
            "brotli-bin-1.0.9     | 19 KB     | : 100% 1.0/1 [00:00<00:00, 34.55it/s]\n",
            "font-ttf-ubuntu-0.83 | 1.9 MB    | : 100% 1.0/1 [00:00<00:00,  3.01it/s]\n",
            "libedit-3.1.20191231 | 121 KB    | : 100% 1.0/1 [00:00<00:00, 18.96it/s]\n",
            "krb5-1.19.3          | 1.4 MB    | : 100% 1.0/1 [00:00<00:00,  3.18it/s]\n",
            "unicodedata2-14.0.0  | 498 KB    | : 100% 1.0/1 [00:00<00:00, 10.36it/s]\n",
            "ca-certificates-2022 | 144 KB    | : 100% 1.0/1 [00:00<00:00, 21.22it/s]\n",
            "bwa-0.7.17           | 187 KB    | : 100% 1.0/1 [00:00<00:00, 15.80it/s]\n",
            "mysql-common-8.0.29  | 1.8 MB    | : 100% 1.0/1 [00:00<00:00,  1.82it/s]\n",
            "_openmp_mutex-4.5    | 23 KB     | : 100% 1.0/1 [00:00<00:00, 30.73it/s]\n",
            "pulseaudio-14.0      | 1.7 MB    | : 100% 1.0/1 [00:00<00:00,  1.74it/s]\n",
            "libffi-3.4.2         | 57 KB     | : 100% 1.0/1 [00:00<00:00, 25.08it/s]\n",
            "libnghttp2-1.47.0    | 808 KB    | : 100% 1.0/1 [00:00<00:00,  5.74it/s]\n",
            "libopenblas-0.3.20   | 10.1 MB   | : 100% 1.0/1 [00:01<00:00,  1.96s/it]               \n",
            "libtool-2.4.6        | 511 KB    | : 100% 1.0/1 [00:00<00:00,  6.07it/s]\n",
            "xz-5.2.5             | 343 KB    | : 100% 1.0/1 [00:00<00:00,  8.77it/s]\n",
            "font-ttf-source-code | 684 KB    | : 100% 1.0/1 [00:00<00:00,  6.33it/s]\n",
            "brotli-1.0.9         | 18 KB     | : 100% 1.0/1 [00:00<00:00, 33.47it/s]\n",
            "lerc-3.0             | 216 KB    | : 100% 1.0/1 [00:00<00:00, 15.98it/s]\n",
            "packaging-21.3       | 36 KB     | : 100% 1.0/1 [00:00<00:00, 24.98it/s]\n",
            "libflac-1.3.4        | 474 KB    | : 100% 1.0/1 [00:00<00:00,  9.11it/s]\n",
            "libgcc-ng-12.1.0     | 940 KB    | : 100% 1.0/1 [00:00<00:00,  5.19it/s]\n",
            "libxcb-1.13          | 391 KB    | : 100% 1.0/1 [00:00<00:00,  5.55it/s]\n",
            "libcups-2.3.3        | 4.6 MB    | : 100% 1.0/1 [00:00<00:00,  1.43it/s]\n",
            "libpq-14.3           | 3.0 MB    | : 100% 1.0/1 [00:00<00:00,  1.33it/s]\n",
            "openssl-1.1.1o       | 2.1 MB    | : 100% 1.0/1 [00:00<00:00,  1.99it/s]\n",
            "zlib-1.2.12          | 90 KB     | : 100% 1.0/1 [00:00<00:00, 18.21it/s]\n",
            "gettext-0.19.8.1     | 3.6 MB    | : 100% 1.0/1 [00:01<00:00,  1.06s/it]\n",
            "matplotlib-3.5.2     | 6 KB      | : 100% 1.0/1 [00:00<00:00, 36.87it/s]\n",
            "icu-70.1             | 13.5 MB   | : 100% 1.0/1 [00:02<00:00,  2.35s/it]               \n",
            "matplotlib-base-3.5. | 7.3 MB    | : 100% 1.0/1 [00:01<00:00,  1.58s/it]\n",
            "libclang13-14.0.4    | 10.6 MB   | : 100% 1.0/1 [00:02<00:00,  2.39s/it]               \n",
            "certifi-2022.5.18.1  | 150 KB    | : 100% 1.0/1 [00:00<00:00, 17.18it/s]\n",
            "pip-22.1.2           | 1.5 MB    | : 100% 1.0/1 [00:00<00:00,  2.05it/s]\n",
            "tk-8.6.12            | 3.3 MB    | : 100% 1.0/1 [00:00<00:00,  1.43it/s]\n",
            "tornado-6.1          | 649 KB    | : 100% 1.0/1 [00:00<00:00,  4.59it/s]\n",
            "pcre-8.45            | 253 KB    | : 100% 1.0/1 [00:00<00:00,  8.97it/s]\n",
            "libwebp-1.2.2        | 85 KB     | : 100% 1.0/1 [00:00<00:00, 18.66it/s]\n",
            "libudev1-249         | 109 KB    | : 100% 1.0/1 [00:00<00:00, 25.57it/s]\n",
            "libuuid-2.32.1       | 28 KB     | : 100% 1.0/1 [00:00<00:00, 31.70it/s]\n",
            "cycler-0.11.0        | 10 KB     | : 100% 1.0/1 [00:00<00:00, 36.87it/s]\n",
            "giflib-5.2.1         | 77 KB     | : 100% 1.0/1 [00:00<00:00, 22.88it/s]\n",
            "fontconfig-2.14.0    | 305 KB    | : 100% 1.0/1 [00:00<00:00,  8.31it/s]\n",
            "libbrotlicommon-1.0. | 65 KB     | : 100% 1.0/1 [00:00<00:00, 25.70it/s]\n",
            "qt-main-5.15.3       | 62.0 MB   | : 100% 1.0/1 [00:14<00:00, 14.76s/it]               \n",
            "zstd-1.5.2           | 452 KB    | : 100% 1.0/1 [00:00<00:00,  9.03it/s]\n",
            "libnsl-2.0.0         | 31 KB     | : 100% 1.0/1 [00:00<00:00, 29.88it/s]\n",
            "libcurl-7.83.1       | 342 KB    | : 100% 1.0/1 [00:00<00:00,  6.28it/s]\n",
            "pyqt5-sip-12.9.0     | 84 KB     | : 100% 1.0/1 [00:00<00:00, 20.59it/s]\n",
            "bedtools-2.30.0      | 15.5 MB   | : 100% 1.0/1 [00:03<00:00,  3.76s/it]\n",
            "libev-4.33           | 104 KB    | : 100% 1.0/1 [00:00<00:00, 21.79it/s]\n",
            "nss-3.78             | 2.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.23it/s]\n",
            "jpeg-9e              | 268 KB    | : 100% 1.0/1 [00:00<00:00, 12.61it/s]\n",
            "pytz-2022.1          | 242 KB    | : 100% 1.0/1 [00:00<00:00,  7.16it/s]\n",
            "xorg-libxdmcp-1.1.3  | 19 KB     | : 100% 1.0/1 [00:00<00:00, 26.39it/s]\n",
            "python-3.9.13        | 26.7 MB   | : 100% 1.0/1 [00:05<00:00,  5.08s/it]               \n",
            "fonts-conda-ecosyste | 4 KB      | : 100% 1.0/1 [00:00<00:00, 33.75it/s]\n",
            "c-ares-1.18.1        | 113 KB    | : 100% 1.0/1 [00:00<00:00, 19.76it/s]\n",
            "libbrotlienc-1.0.9   | 287 KB    | : 100% 1.0/1 [00:00<00:00, 14.84it/s]\n",
            "scipy-1.8.1          | 24.9 MB   | : 100% 1.0/1 [00:05<00:00,  5.38s/it]               \n",
            "cython-0.29.30       | 2.1 MB    | : 100% 1.0/1 [00:00<00:00,  1.58it/s]\n",
            "libcap-2.64          | 96 KB     | : 100% 1.0/1 [00:00<00:00, 16.71it/s]\n",
            "ld_impl_linux-64-2.3 | 667 KB    | : 100% 1.0/1 [00:00<00:00,  6.03it/s]\n",
            "pillow-9.1.1         | 44.5 MB   | : 100% 1.0/1 [00:05<00:00,  5.97s/it]\n",
            "star-2.7.10a         | 4.9 MB    | : 100% 1.0/1 [00:00<00:00,  1.10it/s]\n",
            "perl-5.32.1          | 14.4 MB   | : 100% 1.0/1 [00:03<00:00,  3.92s/it]               \n",
            "libopus-1.3.1        | 255 KB    | : 100% 1.0/1 [00:00<00:00, 12.28it/s]\n",
            "toml-0.10.2          | 18 KB     | : 100% 1.0/1 [00:00<00:00, 26.40it/s]\n",
            "nspr-4.32            | 233 KB    | : 100% 1.0/1 [00:00<00:00,  9.54it/s]\n",
            "libgfortran5-12.1.0  | 1.8 MB    | : 100% 1.0/1 [00:00<00:00,  2.60it/s]\n",
            "libbrotlidec-1.0.9   | 33 KB     | : 100% 1.0/1 [00:00<00:00, 30.73it/s]\n",
            "six-1.16.0           | 14 KB     | : 100% 1.0/1 [00:00<00:00, 29.44it/s]\n",
            "libiconv-1.16        | 1.4 MB    | : 100% 1.0/1 [00:00<00:00,  4.25it/s]\n",
            "lcms2-2.12           | 443 KB    | : 100% 1.0/1 [00:00<00:00,  9.02it/s]\n",
            "numpy-1.22.4         | 6.8 MB    | : 100% 1.0/1 [00:01<00:00,  1.63s/it]\n",
            "_libgcc_mutex-0.1    | 3 KB      | : 100% 1.0/1 [00:00<00:00, 36.00it/s]\n",
            "kiwisolver-1.4.2     | 75 KB     | : 100% 1.0/1 [00:00<00:00, 25.47it/s]\n",
            "font-ttf-dejavu-sans | 388 KB    | : 100% 1.0/1 [00:00<00:00, 10.96it/s]\n",
            "libllvm14-14.0.4     | 35.2 MB   | : 100% 1.0/1 [00:06<00:00,  6.74s/it]               \n",
            "libgfortran-ng-12.1. | 23 KB     | : 100% 1.0/1 [00:00<00:00, 20.88it/s]\n",
            "tzdata-2022a         | 121 KB    | : 100% 1.0/1 [00:00<00:00,  7.95it/s]\n",
            "libxkbcommon-1.0.3   | 581 KB    | : 100% 1.0/1 [00:00<00:00,  6.73it/s]\n",
            "wheel-0.37.1         | 31 KB     | : 100% 1.0/1 [00:00<00:00, 22.56it/s]\n",
            "libclang-14.0.4      | 127 KB    | : 100% 1.0/1 [00:00<00:00, 16.55it/s]\n",
            "lz4-c-1.9.3          | 179 KB    | : 100% 1.0/1 [00:00<00:00, 15.36it/s]\n",
            "python_abi-3.9       | 4 KB      | : 100% 1.0/1 [00:00<00:00, 34.28it/s]\n",
            "pysam-0.19.1         | 2.7 MB    | : 100% 1.0/1 [00:00<00:00,  1.84it/s]\n",
            "alsa-lib-1.2.3.2     | 554 KB    | : 100% 1.0/1 [00:00<00:00,  6.23it/s]\n",
            "pandas-1.4.2         | 12.5 MB   | : 100% 1.0/1 [00:03<00:00,  3.23s/it]               \n",
            "python-dateutil-2.8. | 240 KB    | : 100% 1.0/1 [00:00<00:00, 13.96it/s]\n",
            "libogg-1.3.4         | 206 KB    | : 100% 1.0/1 [00:00<00:00,  9.83it/s]\n",
            "openjpeg-2.4.0       | 444 KB    | : 100% 1.0/1 [00:00<00:00,  6.54it/s]\n",
            "sip-6.5.1            | 389 KB    | : 100% 1.0/1 [00:00<00:00,  7.19it/s]\n",
            "fonts-conda-forge-1  | 4 KB      | : 100% 1.0/1 [00:00<00:00, 32.47it/s]\n",
            "libgomp-12.1.0       | 459 KB    | : 100% 1.0/1 [00:00<00:00,  9.80it/s]\n",
            "libpng-1.6.37        | 306 KB    | : 100% 1.0/1 [00:00<00:00, 11.13it/s]\n",
            "keyutils-1.6.1       | 115 KB    | : 100% 1.0/1 [00:00<00:00, 17.74it/s]\n",
            "pyparsing-3.0.9      | 79 KB     | : 100% 1.0/1 [00:00<00:00, 22.90it/s]\n",
            "ncurses-6.3          | 1002 KB   | : 100% 1.0/1 [00:00<00:00,  1.89it/s]\n",
            "xorg-libxau-1.0.9    | 13 KB     | : 100% 1.0/1 [00:00<00:00, 16.13it/s]\n",
            "freetype-2.10.4      | 890 KB    | : 100% 1.0/1 [00:00<00:00,  4.84it/s]\n",
            "fonttools-4.33.3     | 1.6 MB    | : 100% 1.0/1 [00:00<00:00,  2.03it/s]\n",
            "Preparing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "Verifying transaction: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "Executing transaction: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "#\n",
            "# To activate this environment, use\n",
            "#\n",
            "#     $ conda activate lab\n",
            "#\n",
            "# To deactivate an active environment, use\n",
            "#\n",
            "#     $ conda deactivate\n",
            "\n",
            "Activated conda environment `lab'!\n"
          ]
        }
      ],
      "source": [
        "!git clone https://github.com/hyeshik/colab-biolab.git\n",
        "!cd colab-biolab && bash tools/setup.sh\n",
        "exec(open('colab-biolab/tools/activate_conda.py').read())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "neteyWsWEKBy",
        "outputId": "cef1b526-62d2-4e19-d386-4636529ba9a8",
        "vscode": {
          "languageId": "python"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Solving environment: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.12.0\n",
            "  latest version: 4.13.0\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /root/conda/envs/lab\n",
            "\n",
            "  added / updated specs:\n",
            "    - bedtools\n",
            "    - bioawk\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    bioawk-1.0                 |       h7132678_7         196 KB  bioconda\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         196 KB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  bioawk             bioconda/linux-64::bioawk-1.0-h7132678_7\n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "bioawk-1.0           | 196 KB    | : 100% 1.0/1 [00:00<00:00, 10.16it/s]\n",
            "Preparing transaction: \\ \b\bdone\n",
            "Verifying transaction: / \b\bdone\n",
            "Executing transaction: \\ \b\bdone\n"
          ]
        }
      ],
      "source": [
        "!conda install -y bedtools bioawk"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IANz6DmLUkA6",
        "outputId": "4547d03f-46af-4d80-875a-a01d6521545b",
        "vscode": {
          "languageId": "python"
        }
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/drive/MyDrive/binfo1-datapack01\n"
          ]
        }
      ],
      "source": [
        "%cd /content/drive/MyDrive/binfo1-datapack01/"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### \bFig SC3 그리기 \n",
        "\n",
        "전체 Genome에 대해서 pileup 해주기 "
      ],
      "metadata": {
        "id": "VSq9i7PcleTa"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!samtools mpileup CLIP-35L33G.bam > CLIP-35L33G.pileup\n",
        "!wc -l CLIP-35L33G.pileup"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NozmmYYqlbtB",
        "outputId": "c9d08227-cc6c-42c8-bd39-3406e5119136"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "[mpileup] 1 samples in 1 input files\n",
            "959697300 CLIP-35L33G.pileup\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!head CLIP-35L33G.pileup"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "D1oYh4a8pa01",
        "outputId": "94b08f6b-06f2-41fe-dfee-82573fc61b02"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "chr1\t3056473\tN\t4\t^!G^!G^!G^!G\tHGGI\n",
            "chr1\t3056474\tN\t4\tAAAA\tHB@I\n",
            "chr1\t3056475\tN\t4\tAAAA\tGG?I\n",
            "chr1\t3056476\tN\t4\tTTTT\tHGGH\n",
            "chr1\t3056477\tN\t4\tGGGG\tHEGI\n",
            "chr1\t3056478\tN\t4\tGGGG\tGEEI\n",
            "chr1\t3056479\tN\t4\tAAAA\tEA=I\n",
            "chr1\t3056480\tN\t4\tAAAA\t@AEH\n",
            "chr1\t3056481\tN\t4\tGGGG\tGFFI\n",
            "chr1\t3056482\tN\t4\tTTTT\tG=BG\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!tail CLIP-35L33G.pileup"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dD5xViVEv5KM",
        "outputId": "ba0da1cf-d037-4c32-df09-9052d885e2ac"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "MU069435.1\t31102\tN\t4\tGGGG\tIDDB\n",
            "MU069435.1\t31103\tN\t4\tGGGG\tDEBE\n",
            "MU069435.1\t31104\tN\t4\tGGGG\tEDDE\n",
            "MU069435.1\t31105\tN\t4\tTTTT\t@><8\n",
            "MU069435.1\t31106\tN\t4\tTTT$T$\tDG<;\n",
            "MU069435.1\t31107\tN\t2\tGG\tBC\n",
            "MU069435.1\t31108\tN\t2\tGG\tB>\n",
            "MU069435.1\t31109\tN\t2\tTT\t@3\n",
            "MU069435.1\t31110\tN\t2\tGG\tB1\n",
            "MU069435.1\t31111\tN\t2\tG$G$\t@6\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Filtering with depth (30)"
      ],
      "metadata": {
        "id": "ApQljVIR0xqO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!awk '$4 > 30 {print $0;}' CLIP-35L33G.pileup > CLIP-35L33G-filt.pileup\n",
        "!wc -l CLIP-35L33G-filt.pileup"
      ],
      "metadata": {
        "id": "H47hbqmVwBiI",
        "outputId": "f7d3a7d9-b056-42b7-92cf-e2fa680afaff",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "45781030 CLIP-35L33G-filt.pileup\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!tail CLIP-35L33G-filt.pileup"
      ],
      "metadata": {
        "id": "cUd_8fr2dZQA",
        "outputId": "32d522d5-0ec2-4b22-94cd-9d3fc4088fca",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "MU069435.1\t1559\tN\t346\t>>>>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT$TT$T$T$TTTTTTTTTTTTTTT$TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\tDIDHGHIIIIIIEEIEIGIIEGGBGGDG:BEBHDDFBIGIEHGIIGIIIHBGIBB>G?DI3IGBEI7GGGGI4GII=G?IIGDIIHIIEHGIHIHGEIHIG=EIHGIIHIGIIIIDIDIIIEIIIIIIGIHIIFI@GGIIIIBIIEIGHH?@IAIGI@HIBHHHFAGIGHGHDIIGIGIIBGIEFHIGGIIIIGHIHIDIIFIHIGGIH@DHGHHDIEDIIEIII8GDHHGHIBGIIHIDIAIGIHIIHEG4II>HD?IH@IFIHGIGDHH;AFGH@EGH:BCHG>IEIDIHIHIIHIIIIIIHHGHIGIIGIGFHGCII=HD@HI?IIIGHGGIHIGIDDH>HFB\n",
            "MU069435.1\t1560\tN\t341\t>>>>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT$T$T$T$TT$T$TT$T$T$T$TT$TT$TT$TTT$TT$T$T$TT$TT$T$TTT$T$TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\tDIDHGHIIIIIIDGIEIGIIEGGBGGDG5BEBHEDFGIGIGHHIIGIIIH>GI@DIBIH=HI7HEHGI5II=GBIIDGIIHIGGHBHHIHGGIHIEEGIHGIIIIIIHIIDIBIIIEIIIEIIHIHIIDHCEHIIIIEIIDIGHHBBIAIGI@HIBHHICAGIGHEHBIIGIGIIGDIGDHIGGIHHIGHIAIDGIHIHIIGIH@DHFHHHIGEHIBIIIEHGHGGHIBEIIIIGIBIEDIIIHEHBII?GEBIHDGHIIGIGGHH=GGHHBDGGBHCFHGIBIEIGIHIIHIIII@EHHGHIGIIGIEGHG@II=DD>EI?HIHGHIGGHIGIGDH<HGD\n",
            "MU069435.1\t1561\tN\t317\t>>>>AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\tDIDHEHIGIHDHEEI<EGIH<GG??G;DB;E5BHHHG<><CID@EIH<GEFI@GHDHIIGHBIHIDGBHHHGAGIHGIIIIHHIDGDI@III=IGGBIGGIBIGFHGEH?IIIGIIFHGEH?GI@IDI;HIBGEGHEGGDGGHDIHGFGIIBBIGFHIBGHIIIBHIGIDEGDGHFIEIHEDGBHDGIDEII3IIIBDBDEBDI=BIIIIDI:GHHIIIH4HBIFBH:4GHDFHIH<IDGGH;EGBGDEDH?G@IHGGGIGGHIHIIGIIIIDBHGIHEBIIDIGDFBCII=HGHDI?IIIDDIDIHIIIB@HDEGD\n",
            "MU069435.1\t1562\tN\t319\t>>>>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT$TT$T$TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\tDIDHDHIIIIIIEDIFIGIIEGGBGG@E=:A7G;BIGIHGBD:G>IGII=G?IIGGIIHIHGHGIIIHEDIHIBBGIGGIIIIGIHIIDIDIIIEIIIIIGHIGIIFI@GHIIII:IIGIGGH;DIBIDI@HI?GHIH@GIDHGBDIIGIEIIGEIGFHIGEIIHIGHIGHDHIHIGIGDIHEBHGHHGIG@IIEIII?HDHHBIIBGIIIIGHAIHEIIIHDH:II;HEBIG8IHIIGHGGHH;BBHHD<GHBHCIHBHIIDIHIHIIHIIIIEIF>IHIGIIGIGEHGCII=HDHHI=IIIDGIEIHIIIG@DBHBD\n",
            "MU069435.1\t1563\tN\t316\t>>>>AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA$A$A$A$A$AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA$AAAAAAAAAAAAAA$AAAAAA$AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA$AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA$AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA$AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\tDIDHGFIHIIFI>EHDIGIIE4G;GD=E?7;1E8EIHIHD@8@4II8DEIGGGIGHIIDB>IIIHDDI@IDEGGHBHIIIDIIIHDIDIIIGIIIIIDDIDGIDIHGHBHIHGIIBIDHH4GIHHDG=HH?HHGHHGIBHGBDIIGHGIIGBIG<DIGGIIIIGBIHI:HIDHHIHBBHHDEGHDDIGDHIAIIG?HDHHBII?EIIIIGI=IGIIIIH:G?II>E?;IG1IHGGDID<GB4D>ED3><BGG@IHEIFGGHHIHIIEIIIIIIHGIHIGIIHIHGHH@IIDB?GHIDHIGEGIBHEIGIE4HBHBD\n",
            "MU069435.1\t1564\tN\t305\t>>>>TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT$TTTTTTTTTTTTTTTTTTTTTTTTTTTTT$TTTTTTTTTTTTTTTTTTTT$TTT$TTTTTTTTTT$TT$TTTTTTTTTT$TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\tDIDHGHIIGIGIEDIGIHIIG?FBGG@E=:A7G8@HHIHIIDDBIIGGIHHIHGHDDHHHGGHHIGEHIH@IIGIIIIIIDI7IIIGIIIHIIHIGIDI@HFIIII=IIEDHHDH@IDH@HIEGHGH@EIEHEGEIDGIDIIHEIGGGIGIIGIHHIEI?IIGHGIIGIH@BHHHHGIHDIIGIIGHGHEGIIEDIIIIGI;IHIIIIHDG>II7DE:HH;IHIGIGGGG=GGGHDEEGGG@IAGHIGDIHIHIIGIIIIHGEGIHIGIIHIHEHHCII?GEGHIBIHIDGIEIHIIIBDHBHDB\n",
            "MU069435.1\t1565\tN\t297\t>>>>AAAAAAAAAAAA$AA$AA$A$AA$AA$A$AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\tDIDHBHHIIIGICGIGIEIHDEFBGE8C=1?1G88IHIHIIDGBHIGDIIHIIHHGDIIHEGGGH?BHHHGIIIIIIIIDI:HIIEIGIHIIHIHIDIDDGIIIIIEBHHDHHIGIAIEDBHGGIEGH<IEIGHIGBIGDI@HHIHHHIGI>BIEIHIIBHHFEGHHH?IHDIIDIIDGGHHGIIGGIHIHGIBEHBIHHHDHDIH:G<4IH8BHIHIGDGDDBGGG@GDGGG@IH=IIGGIHIDIIGIIIIIIHGIHFBIIHIHGGHCIIDBFGHI=HGIFGIGIBIGHG?H:BB=\n",
            "MU069435.1\t1566\tN\t289\t>>>>CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCTCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC\tDIDHG>IIHIHHCEIIIEB5D@;=H5BIHIHHI<DAIIBDIIFIIHHGIHIHBDIHI@GHIHBIIIIIIIIDI<IIIEIIIIIHHHHIDIGHIIIGDII>GHH>HGIDIAHBGIHHGI<HGDIGIEIIHFIBHIGIIIIHHI?ICIIHIHIIGIEHEHFHHGIBEIIDIIGHBHFDIIGGIIIIGI@IHDIIIH<FII=HE4HH@IHIIIEE@F7DDGG><EGEGCIH>II2EIHIGIIDIIIIIIHBIHGGGIGIHDHHCII=GD@HIDIIIDGIBIHIIIGBHBHGD\n",
            "MU069435.1\t1567\tN\t291\t>>>>T$T$T$T$T$T$T$T$T$T$T$T$T$T$T$TTTTTTTTTT$T$T$T$T$T$T$T$T$TT$TT$T$T$T$T$T$T$TT$TT$TT$T$T$T$TT$T$T$TT$T$T$TTT$TT$T$TT$T$T$TT$TT$T$T$TTTTTT$TT$T$TTT$TTT$T$T$T$T$TT$T$T$T$T$T$TTT$T$T$T$T$TTT$T$T$T$T$TT$TTT$TTT$T$T$T$T$T$T$T$TT$TT$T$T$T$T$TTT$T$TT$T$T$T$T$T$T$T$T$T$T$T$T$T$T$TT$T$T$T$TTT$T$T$TT$T$T$TT$T$T$T$T$T$T$T$T$T$T$T$T$T$TT$T$T$T$T$T$T$T$T$T$T$TTT$TT$T$T$T$T$T$TT$T$TTT$T$T$TTT$T$T$TT$TTTT$TT$TT$T$T$TTT$TT$T$T$T$T$T$T$TT$TT$T$T$T$T$T$TTT$T$T$T$T$T$T$TT$T$TT$T$T$T$T$T$T$TTT$T$TT$TTT$TTT\tDIDHFHGIIIIIDDIII?B=G::B=G=DIHIHIID@BIIGGIIHIHHHGHII?GDHHIG@HIHGIHIIIIII@IDIIGGDIIIIIHIHIGIDHIIIIEIIGGHH>GGIDI@IEHGHBDHFHHEHGIEIIHDIGHIGIIIIHHIGICIIHIHIIDIHE?HHHHGID4IIGIIDHGHH<II@GIIIIGI:IHIIIIGBH4II:HD7IH0IHIIIGEHH9BGDHD?GBGICIEEIHGGIHIHHIGIIIIGIHGIHIGIIHIHGHHCIIBEEEHI?IIIFGIGIGIIIGBHDHGD\n",
            "MU069435.1\t1568\tN\t88\t>>>>T$T$T$T$T$T$T$T$T$T$TT$T$TT$T$T$T$T$T$T$T$T$T$T$T$TTT$TT$T$T$T$T$T$T$TT$T$T$T$T$TTTT$T$T$T$T$TTT$T$T$T$T$T$T$T$T$TTTTT$T$TT$TT$TT$TT$TT$T$T$T$T$T$T$\tDIDHAG4BDBHBEIGHIHHHIIHIDIIIIIHIGIHIIEIHIHDHIGGIGIIIGIGIIIHIHEG?G@FDIGIIHIEIFG7EIHIDHHGD\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!awk '$1 == \"chr1\" {print $0;}' CLIP-35L33G-filt.pileup > CLIP-33L33G-filt-chr1.pileup"
      ],
      "metadata": {
        "id": "lhhh8QmnLoty"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!head CLIP-33L33G-filt-chr1.pileup"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1cDVbA6ARYUF",
        "outputId": "cf0f701b-486b-46e1-ed17-52c5259b6faa"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "chr1\t3221903\tN\t33\tAAAAAAAAAAAA^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A^!A\tID@@DEG@@IGHEHHDIIIIIGGGBIFGHD@:H\n",
            "chr1\t3221904\tN\t33\tTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\tBDAGDAD;@AHH2HH7IIIIIGHHBIDDGBG@G\n",
            "chr1\t3221905\tN\t33\tTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\tIDAFDDG@@AHD2HEBIIIIEGHH@IEGHGDE?\n",
            "chr1\t3221906\tN\t33\tTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\tAFDH>ED@@AHHDHHDIIIIGGHH@HCEHDBEH\n",
            "chr1\t3221907\tN\t33\tTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\tIFDH@DH@@AFD3HH3IIIIGGGH@HFGHGGEH\n",
            "chr1\t3221908\tN\t33\tAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\tIH@A@<G@=IHH:HHBIIIIIBGHHI@DHBE@G\n",
            "chr1\t3221909\tN\t33\tTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\tHHDH@<G@=FHH=HH<IIIIIGHHCIDAH@9DD\n",
            "chr1\t3221910\tN\t33\tGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG\tIHDH>8D==IHH=HH3IIIIIGHHIIDAHD;DH\n",
            "chr1\t3221911\tN\t33\tAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\tIFDB<BB<BIFB:HHBIIIIIGGHHIGGGG?BH\n",
            "chr1\t3221912\tN\t33\tTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT\tIHDGDBGDDIHHBEH:IIHIIGHHEIDBHGEBH\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Baseread에서 Match와 Substitution만 남기기 "
      ],
      "metadata": {
        "id": "Q3O-HV9V3P_X"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd \n",
        "import re \n",
        "import collections \n",
        "import math\n",
        "\n",
        "def estimate_shannon_entropy(rna_sequence):\n",
        "    m = len(rna_sequence)\n",
        "    bases = collections.Counter([tmp_base for tmp_base in rna_sequence])\n",
        " \n",
        "    shannon_entropy_value = 0\n",
        "    for base in bases:\n",
        "        # number of residues\n",
        "        n_i = bases[base]\n",
        "        # n_i (# residues type i) / M (# residues in column)\n",
        "        p_i = n_i / float(m)\n",
        "        entropy_i = p_i * (math.log(p_i, 2))\n",
        "        shannon_entropy_value += entropy_i\n",
        " \n",
        "    return shannon_entropy_value * -1\n",
        "\n",
        "\n",
        "def chunk_processing(chunck): # define a function that you will use on chunks\n",
        "    toremove = re.compile('-[0-9]+[ACGTNacgtn]+')\n",
        "    chunck['matches'] = chunck['basereads'].apply(lambda x: toremove.sub('', x))\n",
        "    toremove = re.compile('\\+[0-9]+[ACGTNacgtn]+')\n",
        "    chunck['matches'] = chunck['matches'].apply(lambda x: toremove.sub('', x))\n",
        "    toremove = re.compile('[<>$*#]')\n",
        "    chunck['matches'] = chunck['matches'].apply(lambda x: toremove.sub('', x))\n",
        "    toremove = re.compile('[\\^][\\w]')\n",
        "    chunck['matches'] = chunck['matches'].apply(lambda x: toremove.sub('', x))\n",
        "    toremove = re.compile('[\\^][\\W]')\n",
        "    chunck['matches'] = chunck['matches'].apply(lambda x: toremove.sub('', x))\n",
        "    chunck['entropy'] = chunck.apply(lambda x: estimate_shannon_entropy(x['matches']), axis=1)\n",
        "    return chunck\n",
        "\n",
        "\n",
        "chunk_list = [] # create an empty list to hold chunks\n",
        "chunksize = 10 ** 6 # set chunk size\n",
        "for chunk in pd.read_csv('CLIP-35L33G-filt.pileup', sep='\\t', names=['chrom', 'pos', '_ref', 'count', 'basereads', 'quals'],chunksize=chunksize): # read in csv in chunks of chunksize\n",
        "    processed_chunk = chunk_processing(chunk) # process the chunks with chunk_processing() function\n",
        "    chunk_list.append(processed_chunk) # append the chunks to a list\n",
        "df_concat = pd.concat(chunk_list) # concatenate the list to a dataframe\n",
        "df_concat.to_csv(\"CLIP-35L33G-filt.csv\", sep='\\t', index=False)"
      ],
      "metadata": {
        "id": "OR6PAS713ON5"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd \n",
        "import re \n",
        "import collections \n",
        "import math"
      ],
      "metadata": {
        "id": "RNQ7BMp656-C"
      },
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df_concat = pd.read_csv(\"CLIP-35L33G-filt.csv\", sep='\\t')"
      ],
      "metadata": {
        "id": "ApT_gQc-k1hr"
      },
      "execution_count": 6,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Filtering with Entropy (0.7) "
      ],
      "metadata": {
        "id": "yVZ4snC_4Z1L"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "filt_entropy = df_concat[(df_concat['entropy'] >= 1.0) & (df_concat['count'] > 60)].copy()\n",
        "# filt_entropy.to_csv(\"CLIP-35L33G-filt_entropy.csv\", sep='\\t', index=False, columns=['chrom', 'pos', '_ref', 'count', 'basereads', 'quals', 'matches', 'entropy'], )\n",
        "# filt_entropy = pd.read_csv(\"CLIP-35L33G-filt_entropy.csv\", sep='\\t', names=['chrom', 'pos', '_ref', 'count', 'basereads', 'quals', 'entropy'])\n",
        "th_filt_entropy = filt_entropy.sort_values(by='entropy', ascending=False).nlargest(1000, 'entropy')"
      ],
      "metadata": {
        "id": "TIwhAV3q_rDb"
      },
      "execution_count": 9,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "th_filt_entropy"
      ],
      "metadata": {
        "id": "WK0i66hv62om",
        "outputId": "8eac56e1-f233-4442-a14f-9209cc1da20c",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 557
        }
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "          chrom        pos _ref  count  \\\n",
              "45759046   chrY   90803210    N     73   \n",
              "45683317   chrX  168772710    N    168   \n",
              "19109999   chr7   75422105    N     62   \n",
              "30252701  chr11  108902487    N   1046   \n",
              "18304352   chr7   14350331    N    187   \n",
              "...         ...        ...  ...    ...   \n",
              "39531319  chr16   84620886    N    488   \n",
              "18291015   chr7   11601250    N    199   \n",
              "9064642    chr3   89077377    N     95   \n",
              "30186169  chr11  106964921    N    264   \n",
              "1614651    chr1   74319100    N     64   \n",
              "\n",
              "                                                  basereads  \\\n",
              "45759046  >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>><<<<<<<<<<<<<...   \n",
              "45683317  <ccccccccccctcccccttcctcccccccCCCCCccccccccccT...   \n",
              "19109999  <<<<<ggggggggggtattctttctctttggggggtcttttttttc...   \n",
              "30252701  >>>>>>><<TCTCGTAGTGTGCTCCCACGCCTTTTTGGCTTAATTC...   \n",
              "18304352  gggggggggggCCCCTCTTCAGC**CGGCCAG**TTT****CTTCC...   \n",
              "...                                                     ...   \n",
              "39531319  >>>CGGGGGGAAGAAGCGGAAGGAGGCCGAGCAAGGAGGGACGGAG...   \n",
              "18291015  CAGGACGCAGGCAA*GG-2NNCACACAG-2NNAG*G-2NNAGCGGG...   \n",
              "9064642   CCCCCCGGCGGGCCGGCGCTCAGGCCGTGGTCGAGGGAGGGGACGC...   \n",
              "30186169  <tc-2nncccctgtcccttgcggccccgtcttcctc-2nngcgtc-...   \n",
              "1614651   cccccccccaagcacaccaccggacgcggcgccccccccccccccc...   \n",
              "\n",
              "                                                      quals  \\\n",
              "45759046  IIIIEIFDE@GIGI>IIIDIIIG9=I9GIH7CGEIIIHIAIIAGIG...   \n",
              "45683317  G?1IHIGH<DHI@HIDGDGFBGBGHHIHEII3IEGGGHIIGGEIGG...   \n",
              "19109999  EHIIGIIIGIDHEIHIEHDHGIHGGHI@GHHBG=GG?EI>H@EIBI...   \n",
              "30252701  BIHIGD?I4H@<HHBH9@BFEDGGEHIIIIHCIIDGGH3F7FHI;E...   \n",
              "18304352  EHGHIGI:GDD@GGI>H<III@CEDHIBHIIHIIIHI8HDIHHIGD...   \n",
              "...                                                     ...   \n",
              "39531319  HGG6G:BIDIGHBIIHDDHGID7IIIDIBGGIGIHGHHIIH><GEG...   \n",
              "18291015  DI;H;IC;D?HIHEG?IHGDGHF>IGIDCEIIHGDIHCHI0DCIIE...   \n",
              "9064642   FG?G?A?II@GFHIIGIFIHGGHIHGHIGHIIHDGHGHI5I@IH3I...   \n",
              "30186169  /FII@IHIIHH<B>HBGAI?GIIGEGIDG>I<HIHIGIG<>IBIHE...   \n",
              "1614651   AHGGG?7IIGIIGIIHHG5IIIHEGHEIG?4GIGI@:9GG4GBGDI...   \n",
              "\n",
              "                                                    matches   entropy  \n",
              "45759046                           ccTTTTTCCtttttcccccccaaa  2.147317  \n",
              "45683317  ccccccccccctcccccttcctcccccccCCCCCccccccccccTT...  2.107874  \n",
              "19109999  ggggggggggtattctttctctttggggggtcttttttttcccccT...  2.043703  \n",
              "30252701  TCTCGTAGTGTGCTCCCACGCCTTTTTGGCTTAATTCCGTACGGTT...  2.001411  \n",
              "18304352  gggggggggggCCCCTCTTCAGCCGGCCAGTTTCTTCCCTCTCCTC...  1.997098  \n",
              "...                                                     ...       ...  \n",
              "39531319  CGGGGGGAAGAAGCGGAAGGAGGCCGAGCAAGGAGGGACGGAGGGC...  1.627179  \n",
              "18291015  CAGGACGCAGGCAAGGGCAGACCCGCCGCCAGAGAAGAAAGGAAAA...  1.627013  \n",
              "9064642   CCCCCCGGCGGGCCGGCGCTCAGGCCGTGGTCGAGGGAGGGGACGC...  1.626957  \n",
              "30186169           tcctcttccgcttactgctacgcttgtgccccgccccccc  1.626873  \n",
              "1614651   cccccccccaagcacaccaccggacgcggcgccccccccccccccc...  1.626845  \n",
              "\n",
              "[1000 rows x 8 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-c9e792e4-f624-4822-a77f-31f4e219c0d3\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>pos</th>\n",
              "      <th>_ref</th>\n",
              "      <th>count</th>\n",
              "      <th>basereads</th>\n",
              "      <th>quals</th>\n",
              "      <th>matches</th>\n",
              "      <th>entropy</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>45759046</th>\n",
              "      <td>chrY</td>\n",
              "      <td>90803210</td>\n",
              "      <td>N</td>\n",
              "      <td>73</td>\n",
              "      <td>&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;...</td>\n",
              "      <td>IIIIEIFDE@GIGI&gt;IIIDIIIG9=I9GIH7CGEIIIHIAIIAGIG...</td>\n",
              "      <td>ccTTTTTCCtttttcccccccaaa</td>\n",
              "      <td>2.147317</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45683317</th>\n",
              "      <td>chrX</td>\n",
              "      <td>168772710</td>\n",
              "      <td>N</td>\n",
              "      <td>168</td>\n",
              "      <td>&lt;ccccccccccctcccccttcctcccccccCCCCCccccccccccT...</td>\n",
              "      <td>G?1IHIGH&lt;DHI@HIDGDGFBGBGHHIHEII3IEGGGHIIGGEIGG...</td>\n",
              "      <td>ccccccccccctcccccttcctcccccccCCCCCccccccccccTT...</td>\n",
              "      <td>2.107874</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19109999</th>\n",
              "      <td>chr7</td>\n",
              "      <td>75422105</td>\n",
              "      <td>N</td>\n",
              "      <td>62</td>\n",
              "      <td>&lt;&lt;&lt;&lt;&lt;ggggggggggtattctttctctttggggggtcttttttttc...</td>\n",
              "      <td>EHIIGIIIGIDHEIHIEHDHGIHGGHI@GHHBG=GG?EI&gt;H@EIBI...</td>\n",
              "      <td>ggggggggggtattctttctctttggggggtcttttttttcccccT...</td>\n",
              "      <td>2.043703</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30252701</th>\n",
              "      <td>chr11</td>\n",
              "      <td>108902487</td>\n",
              "      <td>N</td>\n",
              "      <td>1046</td>\n",
              "      <td>&gt;&gt;&gt;&gt;&gt;&gt;&gt;&lt;&lt;TCTCGTAGTGTGCTCCCACGCCTTTTTGGCTTAATTC...</td>\n",
              "      <td>BIHIGD?I4H@&lt;HHBH9@BFEDGGEHIIIIHCIIDGGH3F7FHI;E...</td>\n",
              "      <td>TCTCGTAGTGTGCTCCCACGCCTTTTTGGCTTAATTCCGTACGGTT...</td>\n",
              "      <td>2.001411</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18304352</th>\n",
              "      <td>chr7</td>\n",
              "      <td>14350331</td>\n",
              "      <td>N</td>\n",
              "      <td>187</td>\n",
              "      <td>gggggggggggCCCCTCTTCAGC**CGGCCAG**TTT****CTTCC...</td>\n",
              "      <td>EHGHIGI:GDD@GGI&gt;H&lt;III@CEDHIBHIIHIIIHI8HDIHHIGD...</td>\n",
              "      <td>gggggggggggCCCCTCTTCAGCCGGCCAGTTTCTTCCCTCTCCTC...</td>\n",
              "      <td>1.997098</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39531319</th>\n",
              "      <td>chr16</td>\n",
              "      <td>84620886</td>\n",
              "      <td>N</td>\n",
              "      <td>488</td>\n",
              "      <td>&gt;&gt;&gt;CGGGGGGAAGAAGCGGAAGGAGGCCGAGCAAGGAGGGACGGAG...</td>\n",
              "      <td>HGG6G:BIDIGHBIIHDDHGID7IIIDIBGGIGIHGHHIIH&gt;&lt;GEG...</td>\n",
              "      <td>CGGGGGGAAGAAGCGGAAGGAGGCCGAGCAAGGAGGGACGGAGGGC...</td>\n",
              "      <td>1.627179</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18291015</th>\n",
              "      <td>chr7</td>\n",
              "      <td>11601250</td>\n",
              "      <td>N</td>\n",
              "      <td>199</td>\n",
              "      <td>CAGGACGCAGGCAA*GG-2NNCACACAG-2NNAG*G-2NNAGCGGG...</td>\n",
              "      <td>DI;H;IC;D?HIHEG?IHGDGHF&gt;IGIDCEIIHGDIHCHI0DCIIE...</td>\n",
              "      <td>CAGGACGCAGGCAAGGGCAGACCCGCCGCCAGAGAAGAAAGGAAAA...</td>\n",
              "      <td>1.627013</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9064642</th>\n",
              "      <td>chr3</td>\n",
              "      <td>89077377</td>\n",
              "      <td>N</td>\n",
              "      <td>95</td>\n",
              "      <td>CCCCCCGGCGGGCCGGCGCTCAGGCCGTGGTCGAGGGAGGGGACGC...</td>\n",
              "      <td>FG?G?A?II@GFHIIGIFIHGGHIHGHIGHIIHDGHGHI5I@IH3I...</td>\n",
              "      <td>CCCCCCGGCGGGCCGGCGCTCAGGCCGTGGTCGAGGGAGGGGACGC...</td>\n",
              "      <td>1.626957</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30186169</th>\n",
              "      <td>chr11</td>\n",
              "      <td>106964921</td>\n",
              "      <td>N</td>\n",
              "      <td>264</td>\n",
              "      <td>&lt;tc-2nncccctgtcccttgcggccccgtcttcctc-2nngcgtc-...</td>\n",
              "      <td>/FII@IHIIHH&lt;B&gt;HBGAI?GIIGEGIDG&gt;I&lt;HIHIGIG&lt;&gt;IBIHE...</td>\n",
              "      <td>tcctcttccgcttactgctacgcttgtgccccgccccccc</td>\n",
              "      <td>1.626873</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1614651</th>\n",
              "      <td>chr1</td>\n",
              "      <td>74319100</td>\n",
              "      <td>N</td>\n",
              "      <td>64</td>\n",
              "      <td>cccccccccaagcacaccaccggacgcggcgccccccccccccccc...</td>\n",
              "      <td>AHGGG?7IIGIIGIIHHG5IIIHEGHEIG?4GIGI@:9GG4GBGDI...</td>\n",
              "      <td>cccccccccaagcacaccaccggacgcggcgccccccccccccccc...</td>\n",
              "      <td>1.626845</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1000 rows × 8 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-c9e792e4-f624-4822-a77f-31f4e219c0d3')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-c9e792e4-f624-4822-a77f-31f4e219c0d3 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-c9e792e4-f624-4822-a77f-31f4e219c0d3');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "sorting 결과에서 앞뒤로 각 10nt씩 21nt단위로 fasta file 만들기 \n"
      ],
      "metadata": {
        "id": "G5W06LbvKhLa"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "(-)strand에 cross-linked된 read를 고려하기위해 annotation file load"
      ],
      "metadata": {
        "id": "eEDRt-ooGAVD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install pyranges"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Fk_9wAVsYOgN",
        "outputId": "9e1da09a-9d03-435c-d243-201b4f847cb7"
      },
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting pyranges\n",
            "  Downloading pyranges-0.0.117.tar.gz (684 kB)\n",
            "\u001b[K     |████████████████████████████████| 684 kB 27.3 MB/s \n",
            "\u001b[?25hRequirement already satisfied: cython in /usr/local/lib/python3.7/dist-packages (from pyranges) (0.29.30)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.7/dist-packages (from pyranges) (1.3.5)\n",
            "Collecting ncls>=0.0.63\n",
            "  Downloading ncls-0.0.65.tar.gz (470 kB)\n",
            "\u001b[K     |████████████████████████████████| 470 kB 38.5 MB/s \n",
            "\u001b[?25hRequirement already satisfied: tabulate in /usr/local/lib/python3.7/dist-packages (from pyranges) (0.8.9)\n",
            "Collecting sorted_nearest>=0.0.33\n",
            "  Downloading sorted_nearest-0.0.37.tar.gz (1.3 MB)\n",
            "\u001b[K     |████████████████████████████████| 1.3 MB 47.9 MB/s \n",
            "\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "    Preparing wheel metadata ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting pyrle\n",
            "  Downloading pyrle-0.0.35.tar.gz (425 kB)\n",
            "\u001b[K     |████████████████████████████████| 425 kB 50.1 MB/s \n",
            "\u001b[?25h  Installing build dependencies ... \u001b[?25l\u001b[?25hdone\n",
            "  Getting requirements to build wheel ... \u001b[?25l\u001b[?25hdone\n",
            "    Preparing wheel metadata ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: natsort in /usr/local/lib/python3.7/dist-packages (from pyranges) (5.5.0)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from ncls>=0.0.63->pyranges) (1.21.6)\n",
            "Requirement already satisfied: python-dateutil>=2.7.3 in /usr/local/lib/python3.7/dist-packages (from pandas->pyranges) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2017.3 in /usr/local/lib/python3.7/dist-packages (from pandas->pyranges) (2022.1)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.7.3->pandas->pyranges) (1.15.0)\n",
            "Building wheels for collected packages: pyranges, ncls, sorted-nearest, pyrle\n",
            "  Building wheel for pyranges (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pyranges: filename=pyranges-0.0.117-py3-none-any.whl size=693480 sha256=7c99c983ad57644e2e4837479de8770bac6ac37c46d5cad7efaa545334ed158e\n",
            "  Stored in directory: /root/.cache/pip/wheels/1e/c2/29/66f7e4c0de384a3f525410e00111fb88afeb8fbdde3064526f\n",
            "  Building wheel for ncls (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for ncls: filename=ncls-0.0.65-cp37-cp37m-linux_x86_64.whl size=1750195 sha256=2903896b5c9b6f530f54c8730305e033ab9cceef5a970ca7b448c07a8b561fc2\n",
            "  Stored in directory: /root/.cache/pip/wheels/41/55/3b/f2ba6135fec41fbe60fe7e5ebf7ddc84c55d943a4423d88e52\n",
            "  Building wheel for sorted-nearest (PEP 517) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for sorted-nearest: filename=sorted_nearest-0.0.37-cp37-cp37m-linux_x86_64.whl size=4415164 sha256=98be21662d74a49f224bb69849d7ea4773557169aae677aba31dcf9915589ffc\n",
            "  Stored in directory: /root/.cache/pip/wheels/aa/03/91/c3458b46d13e12641eae3658502b10e92288cd44fe9264ff8a\n",
            "  Building wheel for pyrle (PEP 517) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for pyrle: filename=pyrle-0.0.35-cp37-cp37m-linux_x86_64.whl size=1296328 sha256=f7f6b6080b9bbdb2cc586b9a251edd660c66cc435fddcc4907bdc4d342c317f6\n",
            "  Stored in directory: /root/.cache/pip/wheels/3b/ad/e3/9aea7aa6eac18a3c0bc86f72f23f28a51b6d7b2c74b7600f55\n",
            "Successfully built pyranges ncls sorted-nearest pyrle\n",
            "Installing collected packages: sorted-nearest, pyrle, ncls, pyranges\n",
            "Successfully installed ncls-0.0.65 pyranges-0.0.117 pyrle-0.0.35 sorted-nearest-0.0.37\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pyranges as pr\n",
        "\n",
        "# as PyRanges-object\n",
        "# annotation file을 pandas dataframe형태로 변환한다. \n",
        "gr = pr.read_gtf(\"gencode.gtf\")\n",
        "gencode = gr.df"
      ],
      "metadata": {
        "id": "nlQhipUyZOfX"
      },
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "gencode"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 488
        },
        "id": "vvym9Ehckm-z",
        "outputId": "c5e8029a-da31-4e58-f0db-4f3a806e45a9"
      },
      "execution_count": 25,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "        Chromosome   Source      Feature     Start       End Score Strand  \\\n",
              "0             chr1   HAVANA         gene   3143475   3144545     .      +   \n",
              "1             chr1   HAVANA   transcript   3143475   3144545     .      +   \n",
              "2             chr1   HAVANA         exon   3143475   3144545     .      +   \n",
              "3             chr1  ENSEMBL         gene   3172238   3172348     .      +   \n",
              "4             chr1  ENSEMBL   transcript   3172238   3172348     .      +   \n",
              "...            ...      ...          ...       ...       ...   ...    ...   \n",
              "1869102       chrY  ENSEMBL         exon  90850137  90850446     .      -   \n",
              "1869103       chrY  ENSEMBL          CDS  90850140  90850446     .      -   \n",
              "1869104       chrY  ENSEMBL  start_codon  90850443  90850446     .      -   \n",
              "1869105       chrY  ENSEMBL   stop_codon  90850137  90850140     .      -   \n",
              "1869106       chrY  ENSEMBL          UTR  90850137  90850140     .      -   \n",
              "\n",
              "        Frame               gene_id       gene_type  ... transcript_type  \\\n",
              "0           .  ENSMUSG00000102693.2             TEC  ...             NaN   \n",
              "1           .  ENSMUSG00000102693.2             TEC  ...             TEC   \n",
              "2           .  ENSMUSG00000102693.2             TEC  ...             TEC   \n",
              "3           .  ENSMUSG00000064842.3           snRNA  ...             NaN   \n",
              "4           .  ENSMUSG00000064842.3           snRNA  ...           snRNA   \n",
              "...       ...                   ...             ...  ...             ...   \n",
              "1869102     .  ENSMUSG00000096850.2  protein_coding  ...  protein_coding   \n",
              "1869103     0  ENSMUSG00000096850.2  protein_coding  ...  protein_coding   \n",
              "1869104     0  ENSMUSG00000096850.2  protein_coding  ...  protein_coding   \n",
              "1869105     0  ENSMUSG00000096850.2  protein_coding  ...  protein_coding   \n",
              "1869106     .  ENSMUSG00000096850.2  protein_coding  ...  protein_coding   \n",
              "\n",
              "           transcript_name transcript_support_level                 tag  \\\n",
              "0                      NaN                      NaN                 NaN   \n",
              "1        4933401J01Rik-201                       NA               basic   \n",
              "2        4933401J01Rik-201                       NA               basic   \n",
              "3                      NaN                      NaN                 NaN   \n",
              "4              Gm26206-201                       NA               basic   \n",
              "...                    ...                      ...                 ...   \n",
              "1869102        Gm21748-201                       NA  appris_principal_1   \n",
              "1869103        Gm21748-201                       NA  appris_principal_1   \n",
              "1869104        Gm21748-201                       NA  appris_principal_1   \n",
              "1869105        Gm21748-201                       NA  appris_principal_1   \n",
              "1869106        Gm21748-201                       NA  appris_principal_1   \n",
              "\n",
              "            havana_transcript exon_number               exon_id  \\\n",
              "0                         NaN         NaN                   NaN   \n",
              "1        OTTMUST00000127109.1         NaN                   NaN   \n",
              "2        OTTMUST00000127109.1           1  ENSMUSE00001343744.2   \n",
              "3                         NaN         NaN                   NaN   \n",
              "4                         NaN         NaN                   NaN   \n",
              "...                       ...         ...                   ...   \n",
              "1869102                   NaN           1  ENSMUSE00001062768.2   \n",
              "1869103                   NaN           1  ENSMUSE00001062768.2   \n",
              "1869104                   NaN           1  ENSMUSE00001062768.2   \n",
              "1869105                   NaN           1  ENSMUSE00001062768.2   \n",
              "1869106                   NaN           1  ENSMUSE00001062768.2   \n",
              "\n",
              "                   protein_id ccdsid  ont  \n",
              "0                         NaN    NaN  NaN  \n",
              "1                         NaN    NaN  NaN  \n",
              "2                         NaN    NaN  NaN  \n",
              "3                         NaN    NaN  NaN  \n",
              "4                         NaN    NaN  NaN  \n",
              "...                       ...    ...  ...  \n",
              "1869102  ENSMUSP00000137361.2    NaN  NaN  \n",
              "1869103  ENSMUSP00000137361.2    NaN  NaN  \n",
              "1869104  ENSMUSP00000137361.2    NaN  NaN  \n",
              "1869105  ENSMUSP00000137361.2    NaN  NaN  \n",
              "1869106  ENSMUSP00000137361.2    NaN  NaN  \n",
              "\n",
              "[1869107 rows x 25 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-89cfee36-ca50-4cac-8417-16864bd39ed4\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Source</th>\n",
              "      <th>Feature</th>\n",
              "      <th>Start</th>\n",
              "      <th>End</th>\n",
              "      <th>Score</th>\n",
              "      <th>Strand</th>\n",
              "      <th>Frame</th>\n",
              "      <th>gene_id</th>\n",
              "      <th>gene_type</th>\n",
              "      <th>...</th>\n",
              "      <th>transcript_type</th>\n",
              "      <th>transcript_name</th>\n",
              "      <th>transcript_support_level</th>\n",
              "      <th>tag</th>\n",
              "      <th>havana_transcript</th>\n",
              "      <th>exon_number</th>\n",
              "      <th>exon_id</th>\n",
              "      <th>protein_id</th>\n",
              "      <th>ccdsid</th>\n",
              "      <th>ont</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>chr1</td>\n",
              "      <td>HAVANA</td>\n",
              "      <td>gene</td>\n",
              "      <td>3143475</td>\n",
              "      <td>3144545</td>\n",
              "      <td>.</td>\n",
              "      <td>+</td>\n",
              "      <td>.</td>\n",
              "      <td>ENSMUSG00000102693.2</td>\n",
              "      <td>TEC</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>chr1</td>\n",
              "      <td>HAVANA</td>\n",
              "      <td>transcript</td>\n",
              "      <td>3143475</td>\n",
              "      <td>3144545</td>\n",
              "      <td>.</td>\n",
              "      <td>+</td>\n",
              "      <td>.</td>\n",
              "      <td>ENSMUSG00000102693.2</td>\n",
              "      <td>TEC</td>\n",
              "      <td>...</td>\n",
              "      <td>TEC</td>\n",
              "      <td>4933401J01Rik-201</td>\n",
              "      <td>NA</td>\n",
              "      <td>basic</td>\n",
              "      <td>OTTMUST00000127109.1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>HAVANA</td>\n",
              "      <td>exon</td>\n",
              "      <td>3143475</td>\n",
              "      <td>3144545</td>\n",
              "      <td>.</td>\n",
              "      <td>+</td>\n",
              "      <td>.</td>\n",
              "      <td>ENSMUSG00000102693.2</td>\n",
              "      <td>TEC</td>\n",
              "      <td>...</td>\n",
              "      <td>TEC</td>\n",
              "      <td>4933401J01Rik-201</td>\n",
              "      <td>NA</td>\n",
              "      <td>basic</td>\n",
              "      <td>OTTMUST00000127109.1</td>\n",
              "      <td>1</td>\n",
              "      <td>ENSMUSE00001343744.2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>ENSEMBL</td>\n",
              "      <td>gene</td>\n",
              "      <td>3172238</td>\n",
              "      <td>3172348</td>\n",
              "      <td>.</td>\n",
              "      <td>+</td>\n",
              "      <td>.</td>\n",
              "      <td>ENSMUSG00000064842.3</td>\n",
              "      <td>snRNA</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>chr1</td>\n",
              "      <td>ENSEMBL</td>\n",
              "      <td>transcript</td>\n",
              "      <td>3172238</td>\n",
              "      <td>3172348</td>\n",
              "      <td>.</td>\n",
              "      <td>+</td>\n",
              "      <td>.</td>\n",
              "      <td>ENSMUSG00000064842.3</td>\n",
              "      <td>snRNA</td>\n",
              "      <td>...</td>\n",
              "      <td>snRNA</td>\n",
              "      <td>Gm26206-201</td>\n",
              "      <td>NA</td>\n",
              "      <td>basic</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1869102</th>\n",
              "      <td>chrY</td>\n",
              "      <td>ENSEMBL</td>\n",
              "      <td>exon</td>\n",
              "      <td>90850137</td>\n",
              "      <td>90850446</td>\n",
              "      <td>.</td>\n",
              "      <td>-</td>\n",
              "      <td>.</td>\n",
              "      <td>ENSMUSG00000096850.2</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>...</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>Gm21748-201</td>\n",
              "      <td>NA</td>\n",
              "      <td>appris_principal_1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1</td>\n",
              "      <td>ENSMUSE00001062768.2</td>\n",
              "      <td>ENSMUSP00000137361.2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1869103</th>\n",
              "      <td>chrY</td>\n",
              "      <td>ENSEMBL</td>\n",
              "      <td>CDS</td>\n",
              "      <td>90850140</td>\n",
              "      <td>90850446</td>\n",
              "      <td>.</td>\n",
              "      <td>-</td>\n",
              "      <td>0</td>\n",
              "      <td>ENSMUSG00000096850.2</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>...</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>Gm21748-201</td>\n",
              "      <td>NA</td>\n",
              "      <td>appris_principal_1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1</td>\n",
              "      <td>ENSMUSE00001062768.2</td>\n",
              "      <td>ENSMUSP00000137361.2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1869104</th>\n",
              "      <td>chrY</td>\n",
              "      <td>ENSEMBL</td>\n",
              "      <td>start_codon</td>\n",
              "      <td>90850443</td>\n",
              "      <td>90850446</td>\n",
              "      <td>.</td>\n",
              "      <td>-</td>\n",
              "      <td>0</td>\n",
              "      <td>ENSMUSG00000096850.2</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>...</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>Gm21748-201</td>\n",
              "      <td>NA</td>\n",
              "      <td>appris_principal_1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1</td>\n",
              "      <td>ENSMUSE00001062768.2</td>\n",
              "      <td>ENSMUSP00000137361.2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1869105</th>\n",
              "      <td>chrY</td>\n",
              "      <td>ENSEMBL</td>\n",
              "      <td>stop_codon</td>\n",
              "      <td>90850137</td>\n",
              "      <td>90850140</td>\n",
              "      <td>.</td>\n",
              "      <td>-</td>\n",
              "      <td>0</td>\n",
              "      <td>ENSMUSG00000096850.2</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>...</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>Gm21748-201</td>\n",
              "      <td>NA</td>\n",
              "      <td>appris_principal_1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1</td>\n",
              "      <td>ENSMUSE00001062768.2</td>\n",
              "      <td>ENSMUSP00000137361.2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1869106</th>\n",
              "      <td>chrY</td>\n",
              "      <td>ENSEMBL</td>\n",
              "      <td>UTR</td>\n",
              "      <td>90850137</td>\n",
              "      <td>90850140</td>\n",
              "      <td>.</td>\n",
              "      <td>-</td>\n",
              "      <td>.</td>\n",
              "      <td>ENSMUSG00000096850.2</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>...</td>\n",
              "      <td>protein_coding</td>\n",
              "      <td>Gm21748-201</td>\n",
              "      <td>NA</td>\n",
              "      <td>appris_principal_1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1</td>\n",
              "      <td>ENSMUSE00001062768.2</td>\n",
              "      <td>ENSMUSP00000137361.2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>1869107 rows × 25 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-89cfee36-ca50-4cac-8417-16864bd39ed4')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-89cfee36-ca50-4cac-8417-16864bd39ed4 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-89cfee36-ca50-4cac-8417-16864bd39ed4');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 25
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        " # gencode에서 position이 포함되는 gene의 Strand정보를 가져온다. \n",
        " \n",
        " def range_func(x):\n",
        "  try:\n",
        "    return gencode[(gencode['Start'] <= x) & (gencode['End'] >= x)]['Strand'].iloc[0]\n",
        "  except:\n",
        "    return None\n",
        "\n",
        " th_filt_entropy['start'] = th_filt_entropy['pos'] -11\n",
        " th_filt_entropy['end'] = th_filt_entropy['pos'] + 10 \n",
        " th_filt_entropy['strand'] = th_filt_entropy['pos'].apply(lambda x: range_func(x))"
      ],
      "metadata": {
        "id": "oa4Jx2-TVeW8"
      },
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# null 제거 \n",
        "th_filt_entropy = th_filt_entropy[th_filt_entropy['strand'].notnull()]"
      ],
      "metadata": {
        "id": "Q7MCqPcbzHRX"
      },
      "execution_count": 16,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "th_filt_entropy"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 557
        },
        "id": "U-6jMU1t4gik",
        "outputId": "a7f6208d-51d9-46c8-9471-d3ad767f6339"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "          chrom        pos _ref  count  \\\n",
              "45759046   chrY   90803210    N     73   \n",
              "45683317   chrX  168772710    N    168   \n",
              "19109999   chr7   75422105    N     62   \n",
              "30252701  chr11  108902487    N   1046   \n",
              "18304352   chr7   14350331    N    187   \n",
              "...         ...        ...  ...    ...   \n",
              "39531319  chr16   84620886    N    488   \n",
              "18291015   chr7   11601250    N    199   \n",
              "9064642    chr3   89077377    N     95   \n",
              "30186169  chr11  106964921    N    264   \n",
              "1614651    chr1   74319100    N     64   \n",
              "\n",
              "                                                  basereads  \\\n",
              "45759046  >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>><<<<<<<<<<<<<...   \n",
              "45683317  <ccccccccccctcccccttcctcccccccCCCCCccccccccccT...   \n",
              "19109999  <<<<<ggggggggggtattctttctctttggggggtcttttttttc...   \n",
              "30252701  >>>>>>><<TCTCGTAGTGTGCTCCCACGCCTTTTTGGCTTAATTC...   \n",
              "18304352  gggggggggggCCCCTCTTCAGC**CGGCCAG**TTT****CTTCC...   \n",
              "...                                                     ...   \n",
              "39531319  >>>CGGGGGGAAGAAGCGGAAGGAGGCCGAGCAAGGAGGGACGGAG...   \n",
              "18291015  CAGGACGCAGGCAA*GG-2NNCACACAG-2NNAG*G-2NNAGCGGG...   \n",
              "9064642   CCCCCCGGCGGGCCGGCGCTCAGGCCGTGGTCGAGGGAGGGGACGC...   \n",
              "30186169  <tc-2nncccctgtcccttgcggccccgtcttcctc-2nngcgtc-...   \n",
              "1614651   cccccccccaagcacaccaccggacgcggcgccccccccccccccc...   \n",
              "\n",
              "                                                      quals  \\\n",
              "45759046  IIIIEIFDE@GIGI>IIIDIIIG9=I9GIH7CGEIIIHIAIIAGIG...   \n",
              "45683317  G?1IHIGH<DHI@HIDGDGFBGBGHHIHEII3IEGGGHIIGGEIGG...   \n",
              "19109999  EHIIGIIIGIDHEIHIEHDHGIHGGHI@GHHBG=GG?EI>H@EIBI...   \n",
              "30252701  BIHIGD?I4H@<HHBH9@BFEDGGEHIIIIHCIIDGGH3F7FHI;E...   \n",
              "18304352  EHGHIGI:GDD@GGI>H<III@CEDHIBHIIHIIIHI8HDIHHIGD...   \n",
              "...                                                     ...   \n",
              "39531319  HGG6G:BIDIGHBIIHDDHGID7IIIDIBGGIGIHGHHIIH><GEG...   \n",
              "18291015  DI;H;IC;D?HIHEG?IHGDGHF>IGIDCEIIHGDIHCHI0DCIIE...   \n",
              "9064642   FG?G?A?II@GFHIIGIFIHGGHIHGHIGHIIHDGHGHI5I@IH3I...   \n",
              "30186169  /FII@IHIIHH<B>HBGAI?GIIGEGIDG>I<HIHIGIG<>IBIHE...   \n",
              "1614651   AHGGG?7IIGIIGIIHHG5IIIHEGHEIG?4GIGI@:9GG4GBGDI...   \n",
              "\n",
              "                                                    matches   entropy  \\\n",
              "45759046                           ccTTTTTCCtttttcccccccaaa  2.147317   \n",
              "45683317  ccccccccccctcccccttcctcccccccCCCCCccccccccccTT...  2.107874   \n",
              "19109999  ggggggggggtattctttctctttggggggtcttttttttcccccT...  2.043703   \n",
              "30252701  TCTCGTAGTGTGCTCCCACGCCTTTTTGGCTTAATTCCGTACGGTT...  2.001411   \n",
              "18304352  gggggggggggCCCCTCTTCAGCCGGCCAGTTTCTTCCCTCTCCTC...  1.997098   \n",
              "...                                                     ...       ...   \n",
              "39531319  CGGGGGGAAGAAGCGGAAGGAGGCCGAGCAAGGAGGGACGGAGGGC...  1.627179   \n",
              "18291015  CAGGACGCAGGCAAGGGCAGACCCGCCGCCAGAGAAGAAAGGAAAA...  1.627013   \n",
              "9064642   CCCCCCGGCGGGCCGGCGCTCAGGCCGTGGTCGAGGGAGGGGACGC...  1.626957   \n",
              "30186169           tcctcttccgcttactgctacgcttgtgccccgccccccc  1.626873   \n",
              "1614651   cccccccccaagcacaccaccggacgcggcgccccccccccccccc...  1.626845   \n",
              "\n",
              "              start        end strand  \n",
              "45759046   90803199   90803220      +  \n",
              "45683317  168772699  168772720      -  \n",
              "19109999   75422094   75422115      -  \n",
              "30252701  108902476  108902497      -  \n",
              "18304352   14350320   14350341      -  \n",
              "...             ...        ...    ...  \n",
              "39531319   84620875   84620896      -  \n",
              "18291015   11601239   11601260      +  \n",
              "9064642    89077366   89077387      +  \n",
              "30186169  106964910  106964931      -  \n",
              "1614651    74319089   74319110      -  \n",
              "\n",
              "[999 rows x 11 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-aa8d0e09-98cb-43f8-8239-992afe6fac37\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>chrom</th>\n",
              "      <th>pos</th>\n",
              "      <th>_ref</th>\n",
              "      <th>count</th>\n",
              "      <th>basereads</th>\n",
              "      <th>quals</th>\n",
              "      <th>matches</th>\n",
              "      <th>entropy</th>\n",
              "      <th>start</th>\n",
              "      <th>end</th>\n",
              "      <th>strand</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>45759046</th>\n",
              "      <td>chrY</td>\n",
              "      <td>90803210</td>\n",
              "      <td>N</td>\n",
              "      <td>73</td>\n",
              "      <td>&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;...</td>\n",
              "      <td>IIIIEIFDE@GIGI&gt;IIIDIIIG9=I9GIH7CGEIIIHIAIIAGIG...</td>\n",
              "      <td>ccTTTTTCCtttttcccccccaaa</td>\n",
              "      <td>2.147317</td>\n",
              "      <td>90803199</td>\n",
              "      <td>90803220</td>\n",
              "      <td>+</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45683317</th>\n",
              "      <td>chrX</td>\n",
              "      <td>168772710</td>\n",
              "      <td>N</td>\n",
              "      <td>168</td>\n",
              "      <td>&lt;ccccccccccctcccccttcctcccccccCCCCCccccccccccT...</td>\n",
              "      <td>G?1IHIGH&lt;DHI@HIDGDGFBGBGHHIHEII3IEGGGHIIGGEIGG...</td>\n",
              "      <td>ccccccccccctcccccttcctcccccccCCCCCccccccccccTT...</td>\n",
              "      <td>2.107874</td>\n",
              "      <td>168772699</td>\n",
              "      <td>168772720</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19109999</th>\n",
              "      <td>chr7</td>\n",
              "      <td>75422105</td>\n",
              "      <td>N</td>\n",
              "      <td>62</td>\n",
              "      <td>&lt;&lt;&lt;&lt;&lt;ggggggggggtattctttctctttggggggtcttttttttc...</td>\n",
              "      <td>EHIIGIIIGIDHEIHIEHDHGIHGGHI@GHHBG=GG?EI&gt;H@EIBI...</td>\n",
              "      <td>ggggggggggtattctttctctttggggggtcttttttttcccccT...</td>\n",
              "      <td>2.043703</td>\n",
              "      <td>75422094</td>\n",
              "      <td>75422115</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30252701</th>\n",
              "      <td>chr11</td>\n",
              "      <td>108902487</td>\n",
              "      <td>N</td>\n",
              "      <td>1046</td>\n",
              "      <td>&gt;&gt;&gt;&gt;&gt;&gt;&gt;&lt;&lt;TCTCGTAGTGTGCTCCCACGCCTTTTTGGCTTAATTC...</td>\n",
              "      <td>BIHIGD?I4H@&lt;HHBH9@BFEDGGEHIIIIHCIIDGGH3F7FHI;E...</td>\n",
              "      <td>TCTCGTAGTGTGCTCCCACGCCTTTTTGGCTTAATTCCGTACGGTT...</td>\n",
              "      <td>2.001411</td>\n",
              "      <td>108902476</td>\n",
              "      <td>108902497</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18304352</th>\n",
              "      <td>chr7</td>\n",
              "      <td>14350331</td>\n",
              "      <td>N</td>\n",
              "      <td>187</td>\n",
              "      <td>gggggggggggCCCCTCTTCAGC**CGGCCAG**TTT****CTTCC...</td>\n",
              "      <td>EHGHIGI:GDD@GGI&gt;H&lt;III@CEDHIBHIIHIIIHI8HDIHHIGD...</td>\n",
              "      <td>gggggggggggCCCCTCTTCAGCCGGCCAGTTTCTTCCCTCTCCTC...</td>\n",
              "      <td>1.997098</td>\n",
              "      <td>14350320</td>\n",
              "      <td>14350341</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39531319</th>\n",
              "      <td>chr16</td>\n",
              "      <td>84620886</td>\n",
              "      <td>N</td>\n",
              "      <td>488</td>\n",
              "      <td>&gt;&gt;&gt;CGGGGGGAAGAAGCGGAAGGAGGCCGAGCAAGGAGGGACGGAG...</td>\n",
              "      <td>HGG6G:BIDIGHBIIHDDHGID7IIIDIBGGIGIHGHHIIH&gt;&lt;GEG...</td>\n",
              "      <td>CGGGGGGAAGAAGCGGAAGGAGGCCGAGCAAGGAGGGACGGAGGGC...</td>\n",
              "      <td>1.627179</td>\n",
              "      <td>84620875</td>\n",
              "      <td>84620896</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18291015</th>\n",
              "      <td>chr7</td>\n",
              "      <td>11601250</td>\n",
              "      <td>N</td>\n",
              "      <td>199</td>\n",
              "      <td>CAGGACGCAGGCAA*GG-2NNCACACAG-2NNAG*G-2NNAGCGGG...</td>\n",
              "      <td>DI;H;IC;D?HIHEG?IHGDGHF&gt;IGIDCEIIHGDIHCHI0DCIIE...</td>\n",
              "      <td>CAGGACGCAGGCAAGGGCAGACCCGCCGCCAGAGAAGAAAGGAAAA...</td>\n",
              "      <td>1.627013</td>\n",
              "      <td>11601239</td>\n",
              "      <td>11601260</td>\n",
              "      <td>+</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9064642</th>\n",
              "      <td>chr3</td>\n",
              "      <td>89077377</td>\n",
              "      <td>N</td>\n",
              "      <td>95</td>\n",
              "      <td>CCCCCCGGCGGGCCGGCGCTCAGGCCGTGGTCGAGGGAGGGGACGC...</td>\n",
              "      <td>FG?G?A?II@GFHIIGIFIHGGHIHGHIGHIIHDGHGHI5I@IH3I...</td>\n",
              "      <td>CCCCCCGGCGGGCCGGCGCTCAGGCCGTGGTCGAGGGAGGGGACGC...</td>\n",
              "      <td>1.626957</td>\n",
              "      <td>89077366</td>\n",
              "      <td>89077387</td>\n",
              "      <td>+</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30186169</th>\n",
              "      <td>chr11</td>\n",
              "      <td>106964921</td>\n",
              "      <td>N</td>\n",
              "      <td>264</td>\n",
              "      <td>&lt;tc-2nncccctgtcccttgcggccccgtcttcctc-2nngcgtc-...</td>\n",
              "      <td>/FII@IHIIHH&lt;B&gt;HBGAI?GIIGEGIDG&gt;I&lt;HIHIGIG&lt;&gt;IBIHE...</td>\n",
              "      <td>tcctcttccgcttactgctacgcttgtgccccgccccccc</td>\n",
              "      <td>1.626873</td>\n",
              "      <td>106964910</td>\n",
              "      <td>106964931</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1614651</th>\n",
              "      <td>chr1</td>\n",
              "      <td>74319100</td>\n",
              "      <td>N</td>\n",
              "      <td>64</td>\n",
              "      <td>cccccccccaagcacaccaccggacgcggcgccccccccccccccc...</td>\n",
              "      <td>AHGGG?7IIGIIGIIHHG5IIIHEGHEIG?4GIGI@:9GG4GBGDI...</td>\n",
              "      <td>cccccccccaagcacaccaccggacgcggcgccccccccccccccc...</td>\n",
              "      <td>1.626845</td>\n",
              "      <td>74319089</td>\n",
              "      <td>74319110</td>\n",
              "      <td>-</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>999 rows × 11 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-aa8d0e09-98cb-43f8-8239-992afe6fac37')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-aa8d0e09-98cb-43f8-8239-992afe6fac37 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-aa8d0e09-98cb-43f8-8239-992afe6fac37');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Download genome reference"
      ],
      "metadata": {
        "id": "z1Xb6iKAIH_v"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "!wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M29/GRCm39.primary_assembly.genome.fa.gz --no-check-certificate"
      ],
      "metadata": {
        "id": "DezAfxGtuRI7",
        "outputId": "f9749000-4fb6-4d05-90e8-7830df3735ad",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "--2022-06-07 09:09:42--  https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M29/GRCm39.primary_assembly.genome.fa.gz\n",
            "Resolving ftp.ebi.ac.uk (ftp.ebi.ac.uk)... 193.62.193.138\n",
            "Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.193.138|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 773873008 (738M) [application/octet-stream]\n",
            "Saving to: ‘GRCm39.primary_assembly.genome.fa.gz’\n",
            "\n",
            "GRCm39.primary_asse  43%[=======>            ] 319.87M   388KB/s    in 12m 56s \n",
            "\n",
            "2022-06-07 09:22:39 (422 KB/s) - Connection closed at byte 335403944. Retrying.\n",
            "\n",
            "--2022-06-07 09:22:40--  (try: 2)  https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M29/GRCm39.primary_assembly.genome.fa.gz\n",
            "Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.193.138|:443... connected.\n",
            "HTTP request sent, awaiting response... 206 Partial Content\n",
            "Length: 773873008 (738M), 438469064 (418M) remaining [application/octet-stream]\n",
            "Saving to: ‘GRCm39.primary_assembly.genome.fa.gz’\n",
            "\n",
            "GRCm39.primary_asse 100%[++++++++===========>] 738.02M   505KB/s    in 14m 47s \n",
            "\n",
            "2022-06-07 09:37:29 (483 KB/s) - ‘GRCm39.primary_assembly.genome.fa.gz’ saved [773873008/773873008]\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install Bio"
      ],
      "metadata": {
        "id": "1hbfQ1uIt2CN",
        "outputId": "b911e099-7bf0-4926-879f-8f10c015965d",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Collecting Bio\n",
            "  Downloading bio-1.3.9-py3-none-any.whl (270 kB)\n",
            "\u001b[K     |████████████████████████████████| 270 kB 16.9 MB/s \n",
            "\u001b[?25hCollecting biopython>=1.79\n",
            "  Downloading biopython-1.79-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (2.3 MB)\n",
            "\u001b[K     |████████████████████████████████| 2.3 MB 44.8 MB/s \n",
            "\u001b[?25hRequirement already satisfied: tqdm in /usr/local/lib/python3.7/dist-packages (from Bio) (4.64.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.7/dist-packages (from Bio) (2.23.0)\n",
            "Collecting mygene\n",
            "  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython>=1.79->Bio) (1.21.6)\n",
            "Collecting biothings-client>=0.2.6\n",
            "  Downloading biothings_client-0.2.6-py2.py3-none-any.whl (37 kB)\n",
            "Requirement already satisfied: chardet<4,>=3.0.2 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (3.0.4)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2022.5.18.1)\n",
            "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (1.24.3)\n",
            "Requirement already satisfied: idna<3,>=2.5 in /usr/local/lib/python3.7/dist-packages (from requests->Bio) (2.10)\n",
            "Installing collected packages: biothings-client, mygene, biopython, Bio\n",
            "Successfully installed Bio-1.3.9 biopython-1.79 biothings-client-0.2.6 mygene-3.2.2\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO\n",
        "import gzip \n",
        "\n",
        "with gzip.open(\"GRCm39.primary_assembly.genome.fa.gz\", \"rt\") as handle:\n",
        "  seqs = SeqIO.parse(handle, \"fasta\")\n",
        "  with open(\"xlinked_acr1.fasta\", \"a\") as xh:\n",
        "    for seq in seqs:\n",
        "      for idx, row in th_filt_entropy.iterrows():\n",
        "        ## position - 10 to position + 10 \n",
        "        if seq.id == row['chrom']:\n",
        "          xseq = seq.seq[row['start']:row['end']]\n",
        "          print(\"plus\" , str(xseq)[10])\n",
        "          ## (-)strand는 reverse complement seq를 가져온다\n",
        "          if row['strand'] == '-':\n",
        "            xseq = xseq.reverse_complement()\n",
        "            print(\"mius\", str(xseq)[10])\n",
        "          xseq = xseq.transcribe()\n",
        "          # xh.write(str(xseq))\n",
        "          # xh.write(\"\\n\")\n",
        "  xh.close()\n"
      ],
      "metadata": {
        "id": "9b8GAXewqg3y"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "with open(\"xlinked_reverse.fasta\", \"r\") as handle:\n",
        "  lines = handle.readlines()\n",
        "  for line in lines:\n",
        "    "
      ],
      "metadata": {
        "id": "Tdd0evXSmFrY"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        ""
      ],
      "metadata": {
        "id": "DxpP22v31Vj1"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "WebLogo Result "
      ],
      "metadata": {
        "id": "UaJ-VTGhu5ee"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "hexamer 추출하기 "
      ],
      "metadata": {
        "id": "_LtE38mcSQVK"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO\n",
        "import gzip \n",
        "\n",
        "with gzip.open(\"GRCm39.primary_assembly.genome.fa.gz\", \"rt\") as handle:\n",
        "  seqs = SeqIO.parse(handle, \"fasta\")\n",
        "  with open(\"trainset.txt\", \"a\") as xh:\n",
        "    for seq in seqs:\n",
        "      for poskey, posval in pos_dict_.items():\n",
        "        if seq.id == poskey:\n",
        "          for v in posval: \n",
        "            print(v)\n",
        "            print(seq.seq[int(v)-10:int(v)+10])\n",
        "            xh.write(str(seq.seq[int(v)-2:int(v)+4]))\n",
        "            xh.write(\"\\n\")\n",
        "            ## position - 10 to position + 10 \n",
        "  xh.close()"
      ],
      "metadata": {
        "id": "QosCa93tB-xf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "len(filt_entropy)"
      ],
      "metadata": {
        "id": "Pvu2xY5ZDcUh",
        "outputId": "3e07983f-20d0-4fa1-ab7f-eff7f56eea11",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "487484"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "fif_filt_entropy = filt_entropy.nlargest(50000, 'entropy').copy()"
      ],
      "metadata": {
        "id": "aUXI9F7iFzUi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "under_entropy = df_concat[df_concat['entropy'] == 0].sample(50000).copy()"
      ],
      "metadata": {
        "id": "j6Xhr9dLDwU7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "len(under_entropy)"
      ],
      "metadata": {
        "id": "HyU0Q07kD0cH",
        "outputId": "2c993539-7a37-46ac-98cb-52c349a15a83",
        "colab": {
          "base_uri": "https://localhost:8080/"
        }
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "50000"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "under_entropy.to_csv(\"CLIP-35L33G-50-under_entropy.csv\", sep='\\t', index=False, columns=['chrom', 'pos', '_ref', 'count', 'basereads', 'quals', 'entropy'])\n",
        "# fif_filt_entropy.to_csv(\"CLIP-35L33G-50-over_entropy.csv\", sep='\\t', index=False, columns=['chrom', 'pos', '_ref', 'count', 'basereads', 'quals', 'entropy'])"
      ],
      "metadata": {
        "id": "EuHErz0RD3S7"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "fif_filt_entropy['start'] = fif_filt_entropy['pos'] - 2\n",
        "fif_filt_entropy['end'] = fif_filt_entropy['pos'] + 4\n",
        "fif_filt_entropy.to_csv('50-over_entropy.bed', sep='\\t', index=False, header=False, columns=['chrom', 'start', 'end', 'entropy']) "
      ],
      "metadata": {
        "id": "paHRs7hXGbAd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "under_entropy['start'] = under_entropy['pos'] - 2\n",
        "under_entropy['end'] = under_entropy['pos'] + 4\n",
        "under_entropy.to_csv('50-under_entropy.bed', sep='\\t', index=False, header=False, columns=['chrom', 'start', 'end', 'entropy']) "
      ],
      "metadata": {
        "id": "aB1xbNyOFXCj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!bedtools getfasta -fi GRCm39.primary_assembly.genome.fa -bed fifty-over_entropy.bed -fo fifty-over_entropy.fasta\n",
        "!bedtools getfasta -fi GRCm39.primary_assembly.genome.fa -bed fifty-under_entropy.bed -fo fifty-under_entropy.fasta"
      ],
      "metadata": {
        "id": "ANRvJgLvKHFO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!gzip -d GRCm39.primary_assembly.genome.fa.gz"
      ],
      "metadata": {
        "id": "JHWjpUMbL8B5"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Training Model - CNN "
      ],
      "metadata": {
        "id": "LvQdsXQYvBzY"
      }
    }
  ],
  "metadata": {
    "colab": {
      "collapsed_sections": [],
      "name": "CoLab_TermProj_M2.ipynb",
      "provenance": [],
      "toc_visible": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}